---
title: "basicstats"
author: "Jigyasa_Arora"
date: "2/27/2021"
---

##quast results-no.of scaffolds
```{r}
quast<-read.csv("quast_results_allsamples.csv")

print(paste0("shotgun sequencing generated on average= ",mean(quast$all_count), " reads"))

print(paste0("shotgun sequencing generated on average= ",mean(quast$contigs_above1000bps), " scaffolds above 1000bps"))

print(paste0("shotgun sequencing generated on average= ",sum(quast$mapped_count)/sum(quast$all_count)*100, "% of mapped reads"))

```


##samtools results-%mapped reads to GTDB prokaryotic contigs
```{r}
quast<-read.csv("quast_results_allsamples.csv")
mappedreads<-read.csv("prokaryotic_mappedreads.csv")

quast_prokmappedreads<-merge(quast,mappedreads,by="samples")

samples<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv")
samples<-samples%>%select(samples,termite_tree_labels,higher_lower,diet_type_number4)

quastmappedreads_146samples<-merge(quast_prokmappedreads,samples,by="samples")



lower<-quastmappedreads_146samples%>%filter(higher_lower=="lower") #also includes Cryptocercus cockroach
lower2<-quastmappedreads_146samples%>%filter(higher_lower=="lower" & samples!="229-08") #not includes Cryptocercus cockroach
higher<-quastmappedreads_146samples%>%filter(higher_lower=="higher")
wood_higher<-quastmappedreads_146samples%>%filter(diet_type_number4=="2")
soil_higher<-quastmappedreads_146samples%>%filter(diet_type_number4=="3")
fungus_higher<-quastmappedreads_146samples%>%filter(diet_type_number4=="5")

##mean values-
print(paste0("percentage of reads mapped to microbial contigs in all higher termites is ", sum(higher$total_mapped_prok)/sum(higher$all_count)*100))

print(paste0("percentage of reads mapped to microbial contigs in lower termites+Cryptocercus is ",sum(lower$total_mapped_prok)/sum(lower$all_count)*100))

print(paste0("percentage of reads mapped to microbial contigs in lower termites is ",sum(lower2$total_mapped_prok)/sum(lower2$all_count)*100))

print(paste0("percentage of reads mapped to microbial contigs in wood-feeding higher termites is ", sum(wood_higher$total_mapped_prok)/sum(wood_higher$all_count)*100))

print(paste0("percentage of reads mapped to microbial contigs in soil-feeding higher termites is ", sum(soil_higher$total_mapped_prok)/sum(soil_higher$all_count)*100))

print(paste0("percentage of reads mapped to microbial contigs in fungal feeding higher termites is ", sum(fungus_higher$total_mapped_prok)/sum(fungus_higher$all_count)*100))


```


##samtools mapped reads converted to TPM <not needed>
```{r}

##all microbes-
tpmall<-read.csv("tpm_allcontigs_allsamples_allproks.csv")
samples<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv")
samples<-samples%>%select(samples,termite_tree_labels,higher_lower,diet_type_number4)

tpmall_146samples<-merge(tpmall,samples,by="samples")


lower<-tpmall_146samples%>%filter(higher_lower=="lower") #also includes Cryptocercus cockroach
higher<-tpmall_146samples%>%filter(higher_lower=="higher")

print(paste0("percentage of reads (in TPM counts) mapped to microbial contigs in all higher termites is ", sum(higher$prokTPM)/sum(tpmall_146samples$prokTPM)*100))

print(paste0("percentage of reads (in TPM counts) mapped to microbial contigs in lower termites is ",sum(lower$prokTPM)/sum(tpmall_146samples$prokTPM)*100))

#------------------------------------------------------------------------------------
##bacteria only-
tpm<-read.csv("tpm_allcontigs_allsamples_allprokdomains.csv")
tpm<-tpm%>%filter(domain=="Bacteria")


samples<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv")
samples<-samples%>%select(samples,termite_tree_labels,higher_lower,diet_type_number4)

tpm_146samples<-merge(tpm,samples,by="samples")


lower<-tpm_146samples%>%filter(higher_lower=="lower") #also includes Cryptocercus cockroach
higher<-tpm_146samples%>%filter(higher_lower=="higher")

print(paste0("percentage of reads (in TPM counts) mapped to bacterial contigs in all higher termites is ", sum(higher$prokTPM)/sum(tpm_146samples$prokTPM)*100))

print(paste0("percentage of reads (in TPM counts) mapped to bacterial contigs in lower termites is ",sum(lower$prokTPM)/sum(tpm_146samples$prokTPM)*100))

```

