---
title: "functional_taxonomy_cazymes"
author: "Jigyasa_Arora"
date: "2/28/2021"
---

```{r}
library(dplyr)
library(ape)
library(tidyr)
library(plyr)
library(phytools)
library(Information)

```

##Extracting out a)microbes performing CAZYmes in contigs above 5000bps, b) check for duplicate annotations-
```{r}

functions<-read.csv("functions/all-metagenome-annotation-taxonomy-5000bps.txt")
functions$X<-NULL

##extract out CAZYmes-
functions<-functions%>%mutate(databases=ifelse(grepl("K", annotation),"KEGG",ifelse(grepl("hmm",annotation),"CAZY","PFAM")))

functions2<-functions%>%filter(databases=="CAZY") 
write.csv(functions2,file="all-metagenome-annotation-taxonomy-5000bps-cazymes.txt")

#check if there are no repeats in proteinIDs before summing-
functions_prots<-functions2%>%select(fullproteinnames)
functions_prots_uniq<-functions2%>%select(fullproteinnames)%>%unique()

print(paste0("There are no duplicates in CAZYme annotations as ", nrow(functions_prots), "==", nrow(functions_prots_uniq)))

```

##check how many contigs with microbial cazymes have atleast one marker-gene
```{r}
functions2<-read.csv("all-metagenome-annotation-taxonomy-5000bps-cazymes.txt")
colnames(functions2)[which(names(functions2) == "V3")] <- "fullnames"

functions_contigs<-functions2%>%select(fullnames)%>%unique()

markergenes<-read.csv("taxonomy/all-cogs-taxonomy-contiginfo-tpm-filtered-corrected.csv")

#cazyme contigs with markergenes-
functions_contigs_extracted<-functions_contigs%>%filter(fullnames %in% markergenes$fullnames)

print(paste0("Out of ", nrow(functions_contigs), " CAZyme contigs,only ", nrow(functions_contigs_extracted)/nrow(functions_contigs)*100, " have marker-genes" ))
```


##extract cazyme microbial functions part1:from the complete dataset-
```{r}

functions2<-read.csv("all-metagenome-annotation-taxonomy-5000bps-cazymes.txt",row.names=1)

#functions2$V1.y<-gsub("\\|","_",functions2$V1.y)
functions2$phyla<-gsub("_c__.*$","",functions2$taxa)
functions2$class<-gsub("_o__.*$","",functions2$taxa)
functions2$order<-gsub("_f__.*$","",functions2$taxa)
functions2$family<-gsub("_g__.*$","",functions2$taxa)

##get taxa of interest out-
#functions$taxa_of_interest<-paste(functions$domain,functions$phyla,functions$class,functions$family,sep="_")

functions2$taxa_of_interest<-functions2$phyla

##get cazymes out-
library(stringr)
library(data.table)
library(tidyr)

##spread the data- per bacteria per cazyme-
setDT(functions2)
functions_microbegenes<-functions2[, .(TPM = sum(TPM)), by = .(sample_names,annotation,taxa_of_interest)]

functions_microbegenes$annotation<-gsub(".hmm","",functions_microbegenes$annotation)
functions_microbegenes$microbes_cazymes<-paste(functions_microbegenes$annotation,functions_microbegenes$taxa_of_interest,sep="__")

functions_microbegenes<-functions_microbegenes%>%select(sample_names,microbes_cazymes,TPM)
functions_microbegenes_spread<-as.data.frame(spread(functions_microbegenes,microbes_cazymes,TPM))


##extract 129 samples of interest-
tree<-read.tree("1000bps_taxonomy_129taxa_tree.nwk")
d<-data.frame(label=tree$tip.label)
d$runnumber<-unlist(lapply(strsplit(as.character(d$label),split="-"),"[",1))
d$runnumber<-gsub("_","-",d$runnumber)

functions_microbegenes_spread2<-merge(functions_microbegenes_spread,d,by.x="sample_names",by.y="runnumber") #we get 123/129 samples

#order the data-
newtree<-keep.tip(tree,as.vector(functions_microbegenes_spread2$label))

rownames(functions_microbegenes_spread2)<-functions_microbegenes_spread2$label
functions_microbegenes_spread2<-functions_microbegenes_spread2[match(newtree$tip.label,rownames(functions_microbegenes_spread2)), ]

write.csv(functions_microbegenes_spread2,file="cazymes-taxonomy-allcontigs-5000bps_123samples_march2021.csv")
```



##extract cazyme microbial functions part2:from contigs with markergenes-
```{r}
functions2<-read.csv("all-metagenome-annotation-taxonomy-5000bps-cazymes.txt")
colnames(functions2)[which(names(functions2) == "V3")] <- "fullnames"

functions3<-functions2%>%filter(fullnames %in% functions_contigs_extracted$fullnames)

#functions2$V1.y<-gsub("\\|","_",functions2$V1.y)
functions3$phyla<-gsub("_c__.*$","",functions3$taxa)
functions3$class<-gsub("_o__.*$","",functions3$taxa)
functions3$order<-gsub("_f__.*$","",functions3$taxa)
functions3$family<-gsub("_g__.*$","",functions3$taxa)

##get taxa of interest out-
#functions$taxa_of_interest<-paste(functions$domain,functions$phyla,functions$class,functions$family,sep="_")

functions3$taxa_of_interest<-functions3$phyla


library(stringr)
library(data.table)
library(tidyr)

##spread the data- per bacteria per cazyme-
setDT(functions3)
functions_microbegenes<-functions3[, .(TPM = sum(TPM)), by = .(sample_names,annotation,taxa_of_interest)]

functions_microbegenes$annotation<-gsub(".hmm","",functions_microbegenes$annotation)
functions_microbegenes$microbes_cazymes<-paste(functions_microbegenes$annotation,functions_microbegenes$taxa_of_interest,sep="__")

functions_microbegenes<-functions_microbegenes%>%select(sample_names,microbes_cazymes,TPM)
functions_microbegenes_spread<-as.data.frame(spread(functions_microbegenes,microbes_cazymes,TPM))


##extract 129 samples of interest-
tree<-read.tree("1000bps_taxonomy_129taxa_tree.nwk")
d<-data.frame(label=tree$tip.label)
d$runnumber<-unlist(lapply(strsplit(as.character(d$label),split="-"),"[",1))
d$runnumber<-gsub("_","-",d$runnumber)

functions_microbegenes_spread2<-merge(functions_microbegenes_spread,d,by.x="sample_names",by.y="runnumber") #we get 104/124 samples

#sort by new tree-
newtree<-keep.tip(tree,as.vector(functions_microbegenes_spread2$label))

functions_microbegenes_spread2<-functions_microbegenes_spread2[match(newtree$tip.label,functions_microbegenes_spread2$label), ]
rownames(functions_microbegenes_spread2)<-functions_microbegenes_spread2$label

write.csv(functions_microbegenes_spread2,file="cazymes-taxonomy-markergenecontigs-5000bps_123samples_march2021.csv")


```





##transform the data and extract data present in >10% of samples in >3 termite species. both parts here-
```{r}

functions_microbegenes_spread2_part1<-read.csv("cazymes-taxonomy-allcontigs-5000bps_123samples_march2021.csv",row.names=1)
colstoremove<-grep(pattern=".d__Bacteria$",x=names(functions_microbegenes_spread2_part1),value=TRUE)
functions_microbegenes_spread2_part1<-functions_microbegenes_spread2_part1%>%select(-colstoremove)

functions_microbegenes_spread2_part2<-read.csv("cazymes-taxonomy-markergenecontigs-5000bps_123samples_march2021.csv",row.names=1)
colstoremove<-grep(pattern=".d__Bacteria$",x=names(functions_microbegenes_spread2_part2),value=TRUE)
functions_microbegenes_spread2_part2<-functions_microbegenes_spread2_part2%>%select(-colstoremove)


clr_transform<-function(filename_spread2){ 
library(propr)
filename_spread2[is.na(filename_spread2)] <-0 #convert NA to zero
rownames(filename_spread2)<-filename_spread2$label
filename_spread2$label<-NULL
filename_spread2$sample_names<-NULL
#add psuedo counts to account for missing and not-present data-
filename_spread2<-filename_spread2+0.65

filename_propr<-propr(filename_spread2,metric="rho",ivar="clr")

#creating a df with all the transformed pathways and metadata-
filename_propr_df<-filename_propr@logratio

return(filename_propr_df)
}

functions_microbegenes_part1_clr<-clr_transform(functions_microbegenes_spread2_part1)
functions_microbegenes_part2_clr<-clr_transform(functions_microbegenes_spread2_part2)

##extract out GH enzymes only-
library(stringr)
functions_microbegenes_part1_clr_gh<-functions_microbegenes_part1_clr[str_detect(names(functions_microbegenes_part1_clr), "GH")]

functions_microbegenes_part2_clr_gh<-functions_microbegenes_part2_clr[str_detect(names(functions_microbegenes_part2_clr), "GH")]

##filters-
above10perc_function_part1<-function(filename){
filename$label<-NULL 
filename$samplename<-NULL 

filename2<-filename #for filtering
filename2[filename2 > 0] <- 1 
filename2[filename2 <= 0] <-0
filename_10perc<-filename2[colSums(filename2) > 13] 
columns<-colnames(filename_10perc) 

filename3<-filename%>%dplyr::select(columns)
return(filename3)
} #13/123=10%

functions_microbegenes_part1_clr_gh_above10<-above10perc_function_part1(functions_microbegenes_part1_clr_gh)

above10perc_function_part2<-function(filename){
filename$label<-NULL 
filename$samplename<-NULL 

filename2<-filename #for filtering
filename2[filename2 > 0] <- 1 
filename2[filename2 <= 0] <-0
filename_10perc<-filename2[colSums(filename2) > 11] 
columns<-colnames(filename_10perc) 

filename3<-filename%>%dplyr::select(columns)
return(filename3)
} #11/104=10%

functions_microbegenes_part2_clr_gh_above10<-above10perc_function_part1(functions_microbegenes_part2_clr_gh)


morethan3taxa_functions<-function(filename){

filename$label<-NULL
filename$samplename<-NULL
filename2<-filename
filename2[filename2 > 0] <- 1
filename2[filename2 <= 0] <-0
filename_threeabove<-filename2[rowSums(filename2[,])>=3, ]
rows<-rownames(filename_threeabove)

filename3<-subset(filename, rownames(filename) %in% rows)

return(filename3)
} #>3 columns

functions_microbegenes_part1_clr_gh_above10_3above<-morethan3taxa_functions(functions_microbegenes_part1_clr_gh_above10)
functions_microbegenes_part2_clr_gh_above10_3above<-morethan3taxa_functions(functions_microbegenes_part2_clr_gh_above10)

print(paste0("We get ", nrow(functions_microbegenes_part2_clr_gh_above10_3above), " samples, with", ncol(functions_microbegenes_part2_clr_gh_above10_3above), " many columns for contigs with marker genes"  ))

print(paste0("We get ", nrow(functions_microbegenes_part1_clr_gh_above10_3above), " samples, with", ncol(functions_microbegenes_part1_clr_gh_above10_3above), " many columns for contigs with or without marker genes"  ))

##NOTE-We will use the df part1 i.e. with or without markergene data.


write.csv(functions_microbegenes_part1_clr_gh_above10_3above,file="functional-annotation-5000abovecontigs-GHs.csv")

#------------------------------------------------------------------------------------------------
###get the final dataset-
##how many GH enzymes in the final dataset?
ghs<-colnames(functions_microbegenes_part1_clr_gh_above10_3above)%>%as.data.frame()
colnames(ghs)<-c("columns")
ghs$ghnames<-gsub("__d__.*$","",ghs$columns)

print(paste0("We have ", ghs%>%select(ghnames)%>%unique()%>%nrow()," GH enzymes in ", nrow(functions_microbegenes_part1_clr_gh_above10_3above), " termite samples" ))

##how many GH enzymes with >2 microbes associated with it?
ghs_counts<-ghs%>%select(ghnames)%>%group_by(ghnames)%>%count()

print(paste0("We have ", ghs_counts%>%filter(freq>2)%>%nrow()," GH enzymes with more than 1 microbe associated with it "))

##how many GH enzymes have an associated substrate classification?
ghsubstrate<-read.csv("C:/Users/jigyasa-arora.OIST/Dropbox (OIST)/gut bacteria/metagenome analysis/new_stuff/bins/annotation/all_contig_taxonomy/itol_images/version5/lignocellulose_litsearch_substrates.csv")

ghs_counts_susbtrate<-merge(ghs_counts,ghsubstrate,by.x="ghnames",by.y="GH.Family",all.x = TRUE)

print(paste0("There are ",ghs_counts_susbtrate%>%filter(!is.na(Substrate) & freq>2)%>%nrow(), " GHs that have more than 2 microbes associated with them + substrate classification"))

colstoextract<-ghs_counts_susbtrate%>%filter(!is.na(Substrate) & freq>2)%>%select(ghnames)
colstoextract_vector<-paste(colstoextract$ghnames, collapse="__|")

functions_microbegenes_part1_clr_gh_above10_3above_final<-functions_microbegenes_part1_clr_gh_above10_3above[grep(colstoextract_vector, colnames(functions_microbegenes_part1_clr_gh_above10_3above))]

write.csv(functions_microbegenes_part1_clr_gh_above10_3above_final,file="functional-annotation-5000abovecontigs-20GHs.csv")

tree<-read.tree("1000bps_taxonomy_129taxa_tree.nwk")
newtree<-keep.tip(tree,as.vector(rownames(functions_microbegenes_part1_clr_gh_above10_3above_final)))
write.tree(newtree,file="1000bps_taxonomy_110taxa_tree.nwk")
```


## Moran's I and phyl.anova() on microbial groups of each CAZYme-
```{r}

#Moran's I phylogenetic autocorrelation with termite tree-
tree<-read.tree("1000bps_taxonomy_110taxa_tree.nwk")

functions<-read.csv("functional-annotation-5000abovecontigs-20GHs.csv",row.names = 1,header=TRUE)
#functions$X<-NULL
functions<-functions[match(tree$tip.label,rownames(functions)), ] #match with the tree

library(adephylo)
library(ape) 
library(phylobase) 
library(phylosignal) 
 

functions[is.na(functions)] <-0 #convert any NA to zero
functions<-functions[match(tree$tip.label,rownames(functions)), ]
p4d <- phylo4d(tree,functions)
p2<-phyloSignal(p4d = p4d, method = "I")


##method1 for pvalue adjustment-
#adjusted p-value using  fdr method-<p.adjust method>
mod2 = as.data.frame(p2$pvalue)
mod2$padjusted_lambda<-p.adjust(mod2$I, method='fdr')
mod2$functions<-rownames(mod2)
mod2<-mod2%>%dplyr::mutate(significance_padjusted=ifelse(padjusted_lambda<0.05,"*",""))
mod2<-mod2%>%dplyr::mutate(significance_moranI=ifelse(I<0.05,"*",""))
mod2$gh<-gsub("__.*$","",mod2$functions)

genenames<-read.csv("C:/Users/jigyasa-arora.OIST/Dropbox (OIST)/gut bacteria/metagenome analysis/new_stuff/bins/annotation/all_contig_taxonomy/itol_images/version5/lignocellulose_litsearch_substrates.csv")
mod2<-merge(mod2,genenames,by.x="gh",by.y="GH.Family",all.x = TRUE)


write.csv(mod2,file="phylogenetic_autocorrelation_functionalcazymes_110samples_march2021.csv")


##method2 for pvalue adjustment- <doesn't work with small dataset like ours>
#library(fdrtool)
#mod2 = as.data.frame(p2$pvalue)
#mod2pvalue<-fdrtool(as.vector(mod2$I), statistic=c("normal"), plot=FALSE, color.figure=FALSE, #verbose=TRUE,  cutoff.method=c("locfdr"), pct0=0.75)

#mod3<-as.data.frame(mod2pvalue$lfdr)
#rownames(mod3)<-rownames(mod2)
#colnames(mod3)<-c("localfdr")
#mod3<-mod3%>%dplyr::mutate(significance=ifelse(localfdr<0.05,"*",""))
#write.csv(mod3,file="phylogenetic_autocorrelation_cazymes_178columns_124samples_localFDR.csv")
#-----------------------------------------------------------------------------
#Phylogenetic anova
tree<-read.tree("1000bps_taxonomy_110taxa_tree.nwk")

functions<-read.csv("functional-annotation-5000abovecontigs-20GHs.csv",row.names = 1,header=TRUE)
#individually-  running the function in a list of columns-
functions$X<-NULL
functions<-functions[match(tree$tip.label,rownames(functions)), ] #match the tree
functions <-functions+1000 #add a constant to every value

sampleinfo<-read.csv("taxonomy/sample_names_ids_withotherfactors_tomedit.2.csv")

library(geomorph)
phyl.anova.func2_list<-function(filename){
filename<-as.data.frame(t(filename)) #convert row to column
filename$label<-rownames(filename)
geo_functions<-merge(filename,sampleinfo,by.x="label",by.y="termite_tree_labels")
geo_functions$diet_type_number4<-as.factor(geo_functions$diet_type_number4)
geo_functions[,2]<-as.numeric(geo_functions[,2])

gdf <- geomorph.data.frame(abundance = geo_functions[,2], diet = as.factor(geo_functions$diet_type_number4), phy = tree) #create a dataframe that contains the data and phylogeny. [according to google forum]

#group_name <- names(filename)[2] #get the correct column name. column 2 is the cazyme.

geomorph.lm<-procD.lm(abundance ~ diet, iter=9999,RRPP=TRUE,effect.type = "F", SS.type = "II" ,data=gdf,print.progress = FALSE) #SS.type="II" is used if the groups (i.e. diet/lineage) are of uneven sizes.
mod<-anova(geomorph.lm) #to examine if there is a phylogenetic correlation between "y" and "x"
mod2<-as.data.frame(mod$table)
mod2<-mod2[1,7]
# d is distance between vector lengths is the difference in the amount of shape change per unit of size. If d is 0.05 i.e., | 0.20 - 0.15 |.  Why do this?  Two vectors can be correlated in direction - the way shape changes with size - but one group can exhibit more shape change per unit size than the other
#D tells us the effect size of the standard deviation between two groups. Positive and above 0.2-0.3 means larger positive difference, while negative and below -0.2 to -0.3 means larger negative difference.

pw<-pairwise(geomorph.lm,groups = as.factor(gdf$diet))
pw_summary<-summary(pw,confidence=0.95,show.vectors = TRUE)
pw_summary2<-as.data.frame(pw_summary$summary.table)

pw_summary2$padjust<-p.adjust(pw_summary2$`Pr > d`,method="fdr") #fdr correction of p.value
pw_summary2$diets<-rownames(pw_summary2)

#convert the output to correct formats-
names<-colnames(geo_functions[2])
pw_summary2$genename<-rep(names,times=nrow(pw_summary2))
library(tidyr)
pw_summary3<-pw_summary2%>%dplyr::select("padjust","diets","genename")
pw_spread<-pw_summary3%>%spread(diets,padjust)
pw_spread2<-cbind(pw_spread,mod2)

return(pw_spread2)
}

functions_t<-as.data.frame(t(functions))
xy.list <-  split(functions_t, seq(nrow(functions_t))) #convert df to a list

cazymes_geomorph<-lapply(xy.list,phyl.anova.func2_list) #apply on all columns

cazymes_geomorph_df<-do.call(rbind.data.frame,cazymes_geomorph)


#find p.adjust for anova-
cazymes_geomorph_df$padjust<-p.adjust(cazymes_geomorph_df$mod2,method="fdr")
cazymes_geomorph_df<-cazymes_geomorph_df%>%mutate(significant_mod2=ifelse(mod2 <=0.05,"*",""))
cazymes_geomorph_df<-cazymes_geomorph_df%>%mutate(significant_anova=ifelse(padjust <=0.05,"*",""))
cazymes_geomorph_df$gh<-gsub("__.*$","",cazymes_geomorph_df$genename)

genenames<-read.csv("C:/Users/jigyasa-arora.OIST/Dropbox (OIST)/gut bacteria/metagenome analysis/new_stuff/bins/annotation/all_contig_taxonomy/itol_images/version5/lignocellulose_litsearch_substrates.csv") 
cazymes_geomorph_df2<-merge(cazymes_geomorph_df,genenames,by.x="gh",by.y="GH.Family",all.x = TRUE)

write.csv(cazymes_geomorph_df2,file="Phylogenetic_anova_withdiet_functionalcazymes_110samples_march2021.csv")

cazymes_geomorph_df2<-read.csv("Phylogenetic_anova_withdiet_functionalcazymes_110samples_march2021.csv",row.names = 1)
cazymes_geomorph_df2<-cazymes_geomorph_df2%>%mutate(new1.2=ifelse(X1.2<0.05 & X1.2>0.01,"*",ifelse(X1.2<0.01 & X1.2>0.001,"**",ifelse(X1.2<0.001,"***","not-significant"))))
cazymes_geomorph_df2<-cazymes_geomorph_df2%>%mutate(new1.3=ifelse(X1.3<0.05 & X1.3>0.01,"*",ifelse(X1.3<0.01 & X1.3>0.001,"**",ifelse(X1.3<0.001,"***","not-significant"))))
cazymes_geomorph_df2<-cazymes_geomorph_df2%>%mutate(new1.5=ifelse(X1.5<0.05 & X1.5>0.01,"*",ifelse(X1.5<0.01 & X1.5>0.001,"**",ifelse(X1.5<0.001,"***","not-significant"))))
cazymes_geomorph_df2<-cazymes_geomorph_df2%>%mutate(new2.3=ifelse(X2.3<0.05 & X2.3>0.01,"*",ifelse(X2.3<0.01 & X2.3>0.001,"**",ifelse(X2.3<0.001,"***","not-significant"))))
cazymes_geomorph_df2<-cazymes_geomorph_df2%>%mutate(new2.5=ifelse(X2.5<0.05 & X2.5>0.01,"*",ifelse(X2.5<0.01 & X2.5>0.001,"**",ifelse(X2.5<0.001,"***","not-significant"))))
cazymes_geomorph_df2<-cazymes_geomorph_df2%>%mutate(new3.5=ifelse(X3.5<0.05 & X3.5>0.01,"*",ifelse(X3.5<0.01 & X3.5>0.001,"**",ifelse(X3.5<0.001,"***","not-significant"))))

write.csv(cazymes_geomorph_df2,file="Phylogenetic_anova_withdiet_functionalcazymes_110samples_march2021.csv")
```


##relative abundance values- mean TPM for each termite diet <TableS8>-
```{r}



#raw tpm data-
functions_tpm<-read.csv("cazymes-taxonomy-allcontigs-5000bps_123samples_march2021.csv")
functions_tpm[is.na(functions_tpm)] <-0

#clr data-
functions<-read.csv("functional-annotation-5000abovecontigs-20GHs.csv",row.names = 1,header=TRUE)
functions$X<-rownames(functions)

#get the columns and samples equivalent to clr data-
functions_tpm2<-functions_tpm%>%select(colnames(functions))
functions_tpm3<-functions_tpm2%>%filter(X %in% functions$X)

#merge with diet-
sampleinfo<-read.csv("taxonomy/sample_names_ids_withotherfactors_tomedit.2.csv")

functions2<-merge(functions_tpm3,sampleinfo,by.x="X",by.y="termite_tree_labels")

cols<-colnames(functions_tpm3)
cols<-cols[-69]

setDT(functions2)
functions2_diet<-functions2[, lapply(.SD, mean, na.rm=TRUE), by=diet_type_number4, .SDcols=cols ]
functions2_diet<-as.data.frame(functions2_diet)
functions2_diet$diet_type_number4<-paste("diet",functions2_diet$diet_type_number4,sep="_")

functions2_diet2<-t(functions2_diet)%>%as.data.frame()
functions2_diet2$names<-rownames(functions2_diet2)


write.csv(functions2_diet2,file="functional-annotation-5000abovecontigs-phylanova.csv")

```


