---
title: "functions"
author: "Jigyasa_Arora"
date: "2/28/2021"
---

```{r}
library(dplyr)
library(tidyr)
library(data.table)
library(ape)
```

##extracting CAZYme families present in >10% of termite species + clr transformed-
```{r}
above1000<-read.csv("1000bps_genes_feb2021.txt",row.names = 1) #all functional genes TPM
above1000<-above1000%>%mutate(databases=ifelse(grepl("^K", annotation),"KEGG",ifelse(grepl(".hmm",annotation),"CAZY","PFAM")))


above1000%>%select(databases)%>%group_by(databases)%>%count()
#-----------------------------------------------------------
#termite time tree
termite_tree_new<-read.tree("rna12-16S_concensus_may2020_newnames.nwk") #load the tree
dnew<-data.frame(label=termite_tree_new$tip.label)

dnew$label<-as.character(dnew$label)
dnew$runnumber<-unlist(lapply(strsplit(dnew$label,split="-"),"[",1))
dnew$runnumber<-gsub("_","-",dnew$runnumber)

#spread the data-
library(tidyr)
above1000$databases<-NULL
above1000_spread<-as.data.frame(spread(above1000,annotation,TPM))
rownames(above1000_spread)<-above1000_spread$samples

#join the spread file with termite tree label file to get all the sample labels together-
above1000_spread2<-merge(above1000_spread,dnew,by.x="samples",by.y="runnumber",all.y=TRUE)
rownames(above1000_spread2)<-above1000_spread2$label

#order the rownames by termite tree to get samples with no data at this step!
above1000_spread2<-above1000_spread2[match(termite_tree_new$tip.label,rownames(above1000_spread2)), ]

above1000_spread2[is.na(above1000_spread2)] <-0
above1000_spread2$label<-NULL
above1000_spread2$samples<-NULL


#---------------------------------------------------------------------------
#get the 124 samples that were finalized 
newtree_124samples<-read.tree("1000bps_taxonomy_124taxa_tree.nwk")
d<-data.frame(label=newtree_124samples$tip.label)
d$samples<-unlist(lapply(strsplit(as.character(d$label),split="-"),"[",1))
d$samples2<-gsub("_","-",d$samples)

above1000_spread2_final<-subset(above1000_spread2, rownames(above1000_spread2) %in% d$label)
#--------------------------------------------------------------------------
#clr transform the spread data-

clr_transform<-function(filename_spread2){ 
library(propr) 
filename_spread2[is.na(filename_spread2)] <-0 #convert NA to zero
#rownames(filename_spread2)<-filename_spread2$label
filename_spread2$label<-NULL
filename_spread2$samples<-NULL
#add psuedo counts to account for missing and not-present data-  
filename_spread2<-filename_spread2+0.65

filename_propr<-propr(filename_spread2,metric="rho",ivar="clr")

#creating a df with all the transformed pathways and metadata-
filename_propr_df<-filename_propr@logratio

return(filename_propr_df)
}


above1000_functions_clr<-clr_transform(above1000_spread2_final)

##TEST for clr transformation- row sum is zero
round(rowSums(above1000_functions_clr[1,]))

write.csv(above1000_functions_clr,file="functions_124samples_11158columns__clrtransformed_feb2021.csv")
write.csv(above1000_spread2_final,file="functions_124samples_11158columns_rawtpm_feb2021.csv")
#-----------------------------------------------------------

above1000_clr_cazymes<-above1000_functions_clr%>%dplyr::select(matches("(.hmm)")) #get CAZymes out
above1000_tpm_cazymes<-above1000_spread2_final%>%dplyr::select(matches("(.hmm)")) #get CAZymes out
write.csv(above1000_tpm_cazymes,file="cazymes_318columns_124samples_rawTPM_feb2021.csv")

#-----------------------------------------------------------

#from clr transformed data->re-select columns present in 10% of samples (13/124)
above10perc_function<-function(filename){
filename$label<-NULL   
filename$samplename<-NULL   
 
filename2<-filename #for filtering 
filename2[filename2 > 0] <- 1 
filename2[filename2 <= 0] <-0
filename_10perc<-filename2[colSums(filename2) > 13]  
columns<-colnames(filename_10perc) 

filename3<-filename%>%dplyr::select(columns)
return(filename3)
}

#from clr transformed data-> re-select rows with more than 3 functions 
morethan3taxa_functions<-function(filename){
 
filename$label<-NULL 
filename$samplename<-NULL
filename2<-filename
filename2[filename2 > 0] <- 1
filename2[filename2 <= 0] <-0
filename_threeabove<-filename2[rowSums(filename2[,])>=3, ]
rows<-rownames(filename_threeabove)

filename3<-subset(filename, rownames(filename) %in% rows)

return(filename3)
}


#10% of cazymes-
cazymes_clr_10perc<-above10perc_function(above1000_clr_cazymes)
#more than 3 taxa per functions-
cazymes_clr_10perc_3above<-morethan3taxa_functions(cazymes_clr_10perc)


write.csv(cazymes_clr_10perc_3above,file="cazymes_153columns_122samples_clr_feb2021.csv")

newtree_122samples<-keep.tip(newtree_124samples,as.vector(rownames(cazymes_clr_10perc_3above)))
write.tree(newtree_122samples,file="1000bps_taxonomy_122taxa_tree.nwk")

```


##basic stats on CAZymes-relative abundance and overall distribution-
```{r}

##overall distribution of CAZymes in 124 metagenomes -
above1000_tpm_cazymes<-read.csv("cazymes_318columns_124samples_rawTPM_feb2021.csv",row.names = 1)
above1000_tpm_cazymes$X<-rownames(above1000_tpm_cazymes)

#removing zero contaning rows and columns-
above1000_tpm_cazymes_long<-gather(above1000_tpm_cazymes, cazymes, TPM, 1:318, factor_key=TRUE)
above1000_tpm_cazymes_long<-above1000_tpm_cazymes_long%>%filter(TPM>0)

#spread the data back with no zero values-
above1000_tpm_cazymes_nozero<- above1000_tpm_cazymes_long %>% spread(cazymes,TPM)

#extract cazymes present-
cazymes<-as.data.frame(colnames(above1000_tpm_cazymes_nozero))
colnames(cazymes)<-c("annotation")
#cazymes<-cazymes%>%filter(annotation!="SLH.hmm")
cazymes<-cazymes%>%mutate(types=ifelse(grepl("AA", annotation),"AA",ifelse(grepl("GT",annotation),"GT",ifelse(grepl("CBM",annotation),"CBM",ifelse(grepl("GH",annotation),"GH",ifelse(grepl("CE",annotation),"CE","PL"))))))
cazymes<-cazymes%>%filter(annotation!="X")
cazymes%>%select(types)%>%group_by(types)%>%count()

#-----------------------------------------------------------------------------
##total no. of families per metagenomes-
above1000_tpm_cazymes<-read.csv("cazymes_318columns_124samples_rawTPM_feb2021.csv")
above1000_tpm_cazymes_long<-gather(above1000_tpm_cazymes, cazymes, TPM, 2:319, factor_key=TRUE)
above1000_tpm_cazymes_long<-above1000_tpm_cazymes_long%>%filter(TPM>0)

above1000_tpm_cazymes_long_count<-above1000_tpm_cazymes_long%>%select(X)%>%group_by(X)%>%count()
colnames(above1000_tpm_cazymes_long_count)<-c("X","freq")
#write.csv(above1000_tpm_cazymes_long_count,file="cazymes_overallcountpersample_feb2021.csv") 

print(paste0("minimum no.of cazymes in 124 metagenomes are ", min(above1000_tpm_cazymes_long_count$freq)))

print(paste0("maximum no. of cazymes in 124 metagenomes are ", max(above1000_tpm_cazymes_long_count$freq)))

#-----------------------------------------------------------------------
#perc abundance of GH families in metagenomes-

above1000_tpm_cazymes<-read.csv("cazymes_318columns_124samples_rawTPM_feb2021.csv",row.names = 1)
above1000_tpm_cazymes$X<-rownames(above1000_tpm_cazymes)

#removing zero contaning rows and columns-
above1000_tpm_cazymes_long<-gather(above1000_tpm_cazymes, cazymes, TPM, 1:318, factor_key=TRUE)
above1000_tpm_cazymes_long<-above1000_tpm_cazymes_long%>%filter(TPM>0)

#spread the data back with no zero values-
above1000_tpm_cazymes_nozero<- above1000_tpm_cazymes_long %>% spread(cazymes,TPM)
rownames(above1000_tpm_cazymes_nozero)<-above1000_tpm_cazymes_nozero$X
above1000_tpm_cazymes_nozero$X<-NULL

#convert everything to binary-
above1000_tpm_cazymes_nozero[is.na(above1000_tpm_cazymes_nozero)] <-0 #convert NA to zero
above1000_tpm_cazymes_nozero[above1000_tpm_cazymes_nozero > 0] <- 1 
above1000_tpm_cazymes_nozero[above1000_tpm_cazymes_nozero <= 0] <-0

#calculate distribution of cazymes-percolumn sum-
columnsums<-colSums(above1000_tpm_cazymes_nozero)%>%as.data.frame()
colnames(columnsums)<-c("count")
columnsums$total<-rep(123,nrow(columnsums))
columnsums$perc<-columnsums$count/columnsums$total*100
columnsums$cazymes<-rownames(columnsums)


columnsums<-columnsums%>%mutate(types=ifelse(grepl("AA", cazymes),"AA",ifelse(grepl("GT",cazymes),"GT",ifelse(grepl("CBM",cazymes),"CBM",ifelse(grepl("GH",cazymes),"GH",ifelse(grepl("CE",cazymes),"CE","PL"))))))

columnsums%>%filter(types=="GH" & perc >=75)
columnsums%>%filter(types=="GH" & perc <75 & perc >60)


tree<-read.tree("1000bps_taxonomy_124taxa_tree.nwk")
newtree_123samples<-keep.tip(tree,as.vector(rownames(above1000_tpm_cazymes_nozero)))

above1000_tpm_cazymes_nozero<-above1000_tpm_cazymes_nozero[match(newtree_123samples$tip.label,rownames(above1000_tpm_cazymes_nozero)), ]

##NOTE- we can say that CAZymes were acquired early in evolution as 13 GHs present in 60-75% of termite species are also present in the basal termites.
#-----------------------------------------------------------


```





##stats-Moran's I and Phyl.anova()
```{r}

#Moran's I phylogenetic autocorrelation with termite tree-
tree<-read.tree("1000bps_taxonomy_122taxa_tree.nwk")

functions<-read.csv("cazymes_153columns_122samples_clr_feb2021.csv",row.names = 1,header=TRUE)
functions$X<-NULL
functions<-functions[match(tree$tip.label,rownames(functions)), ] #match with the tree

library(adephylo)
library(ape) 
library(phylobase) 
library(phylosignal) 
 

functions[is.na(functions)] <-0 #convert any NA to zero
functions<-functions[match(tree$tip.label,rownames(functions)), ]
p4d <- phylo4d(tree,functions)
p2<-phyloSignal(p4d = p4d, method = "I")


##method1 for pvalue adjustment-
#adjusted p-value using  fdr method-<p.adjust method>
mod2 = as.data.frame(p2$pvalue)
mod2$padjusted_lambda<-p.adjust(mod2$I, method='fdr')
mod2$functions<-rownames(mod2)
mod2<-mod2%>%dplyr::mutate(significance_padjusted=ifelse(padjusted_lambda<0.05,"*",""))
mod2<-mod2%>%dplyr::mutate(significance_moranI=ifelse(I<0.05,"*",""))
mod2$functions<-gsub(".hmm","",mod2$functions)

genenames<-read.csv("C:/Users/jigyasa-arora.OIST/Dropbox (OIST)/gut bacteria/metagenome analysis/new_stuff/bins/annotation/all_contig_taxonomy/itol_images/version2/lignocellulose substrates and GHs lit search.csv")
mod2<-merge(mod2,genenames,by.x="functions",by.y="GH.Family",all.x = TRUE)

write.csv(mod2,file="phylogenetic_autocorrelation_cazymes_153columns_122samples_jan2021.csv")


##method2 for pvalue adjustment- <doesn't work with small dataset like ours>
#library(fdrtool)
#mod2 = as.data.frame(p2$pvalue)
#mod2pvalue<-fdrtool(as.vector(mod2$I), statistic=c("normal"), plot=FALSE, color.figure=FALSE, #verbose=TRUE,  cutoff.method=c("locfdr"), pct0=0.75)

#mod3<-as.data.frame(mod2pvalue$lfdr)
#rownames(mod3)<-rownames(mod2)
#colnames(mod3)<-c("localfdr")
#mod3<-mod3%>%dplyr::mutate(significance=ifelse(localfdr<0.05,"*",""))
#write.csv(mod3,file="phylogenetic_autocorrelation_cazymes_178columns_124samples_localFDR.csv")
#-----------------------------------------------------------------------------
#Phylogenetic anova
tree<-read.tree("1000bps_taxonomy_122taxa_tree.nwk")

functions<-read.csv("cazymes_153columns_122samples_clr_feb2021.csv",row.names = 1,header=TRUE)
#individually-  running the function in a list of columns-
functions$X<-NULL
functions<-functions[match(tree$tip.label,rownames(functions)), ] #match the tree
functions <-functions+1000 #add a constant to every value

sampleinfo<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv")

library(geomorph)
phyl.anova.func2_list<-function(filename){
filename<-as.data.frame(t(filename)) #convert row to column
filename$label<-rownames(filename)
geo_functions<-merge(filename,sampleinfo,by.x="label",by.y="termite_tree_labels")
geo_functions$diet_type_number4<-as.factor(geo_functions$diet_type_number4)
geo_functions[,2]<-as.numeric(geo_functions[,2])

gdf <- geomorph.data.frame(abundance = geo_functions[,2], diet = as.factor(geo_functions$diet_type_number4), phy = tree) #create a dataframe that contains the data and phylogeny. [according to google forum]

#group_name <- names(filename)[2] #get the correct column name. column 2 is the cazyme.

geomorph.lm<-procD.lm(abundance ~ diet, iter=9999,RRPP=TRUE,effect.type = "F", SS.type = "II" ,data=gdf,print.progress = FALSE) #SS.type="II" is used if the groups (i.e. diet/lineage) are of uneven sizes.
mod<-anova(geomorph.lm) #to examine if there is a phylogenetic correlation between "y" and "x"
mod2<-as.data.frame(mod$table)
mod2<-mod2[1,7]
# d is distance between vector lengths is the difference in the amount of shape change per unit of size. If d is 0.05 i.e., | 0.20 - 0.15 |.  Why do this?  Two vectors can be correlated in direction - the way shape changes with size - but one group can exhibit more shape change per unit size than the other
#D tells us the effect size of the standard deviation between two groups. Positive and above 0.2-0.3 means larger positive difference, while negative and below -0.2 to -0.3 means larger negative difference.

pw<-pairwise(geomorph.lm,groups = as.factor(gdf$diet))
pw_summary<-summary(pw,confidence=0.95,show.vectors = TRUE)
pw_summary2<-as.data.frame(pw_summary$summary.table)

pw_summary2$padjust<-p.adjust(pw_summary2$`Pr > d`,method="fdr") #fdr correction of p.value
pw_summary2$diets<-rownames(pw_summary2)

#convert the output to correct formats-
names<-colnames(geo_functions[2])
pw_summary2$genename<-rep(names,times=nrow(pw_summary2))
library(tidyr)
pw_summary3<-pw_summary2%>%dplyr::select("padjust","diets","genename")
pw_spread<-pw_summary3%>%spread(diets,padjust)
pw_spread2<-cbind(pw_spread,mod2)
 
return(pw_spread2) 
}

functions_t<-as.data.frame(t(functions))
xy.list <-  split(functions_t, seq(nrow(functions_t))) #convert df to a list

cazymes_geomorph<-lapply(xy.list,phyl.anova.func2_list) #apply on all columns

cazymes_geomorph_df<-do.call(rbind.data.frame,cazymes_geomorph)


#find p.adjust for anova-
cazymes_geomorph_df$padjust<-p.adjust(cazymes_geomorph_df$mod2,method="fdr")
cazymes_geomorph_df<-cazymes_geomorph_df%>%mutate(significant_mod2=ifelse(mod2 <=0.05,"*",""))
cazymes_geomorph_df<-cazymes_geomorph_df%>%mutate(significant_anova=ifelse(padjust <=0.05,"*",""))
cazymes_geomorph_df$genename<-gsub(".hmm","",cazymes_geomorph_df$genename)

genenames<-read.csv("C:/Users/jigyasa-arora.OIST/Dropbox (OIST)/gut bacteria/metagenome analysis/new_stuff/bins/annotation/all_contig_taxonomy/itol_images/version2/lignocellulose substrates and GHs lit search.csv",header=FALSE) #functional names 
cazymes_geomorph_df2<-merge(cazymes_geomorph_df,genenames,by.x="genename",by.y="V1",all.x = TRUE)

write.csv(cazymes_geomorph_df2,file="Phylogenetic_anova_withdiet_153CAZYmes_122samples_feb2021.csv")

```






##basic stats on cazymes with diet-
```{r} 

#Generatin non-zero data  -
above1000_tpm_cazymes<-read.csv("cazymes_318columns_124samples_rawTPM_feb2021.csv")

#removing zero contaning rows and columns-
above1000_tpm_cazymes_long<-gather(above1000_tpm_cazymes, cazymes, TPM, 2:319, factor_key=TRUE)
above1000_tpm_cazymes_long<-above1000_tpm_cazymes_long%>%filter(TPM>0)

#spread the data back with no zero values-
above1000_tpm_cazymes_nozero<- above1000_tpm_cazymes_long %>% spread(cazymes,TPM)
rownames(above1000_tpm_cazymes_nozero)<-above1000_tpm_cazymes_nozero$X
above1000_tpm_cazymes_nozero$X<-NULL
above1000_tpm_cazymes_nozero[is.na(above1000_tpm_cazymes_nozero)] <-0 #convert NA to zero

#convert everything to binary-
above1000_tpm_cazymes_nozero_binary<-above1000_tpm_cazymes_nozero
above1000_tpm_cazymes_nozero_binary[is.na(above1000_tpm_cazymes_nozero_binary)] <-0 #convert NA to zero
above1000_tpm_cazymes_nozero_binary[above1000_tpm_cazymes_nozero_binary > 0] <- 1 
above1000_tpm_cazymes_nozero_binary[above1000_tpm_cazymes_nozero_binary <= 0] <-0

#-------------------------------------------------------------------------------

## Overall distribution per diet-
above1000_tpm_cazymes_nozero_binary_rows<-above1000_tpm_cazymes_nozero_binary %>% mutate(sumVar = rowSums(.[1:298]))
rownames(above1000_tpm_cazymes_nozero_binary_rows)<-rownames(above1000_tpm_cazymes_nozero_binary)
above1000_tpm_cazymes_nozero_binary_rows<-above1000_tpm_cazymes_nozero_binary_rows%>%select(sumVar)

above1000_tpm_cazymes_nozero_binary_rows$X<-rownames(above1000_tpm_cazymes_nozero_binary_rows)

sampleinfo<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv")
above1000_tpm_cazymes_nozero_binary_rows<-merge(above1000_tpm_cazymes_nozero_binary_rows,sampleinfo,by.x="X",by.y="termite_tree_labels")

setDT(above1000_tpm_cazymes_nozero_binary_rows)
binarycazymes_diet<-above1000_tpm_cazymes_nozero_binary_rows[, .(sumVar = sum(sumVar)), by = .(diet_type_number4)]


head(binarycazymes_diet,5)
#--------------------------------------------------------------------------------
#relative abundance of cazymes by diet-
above1000_tpm_cazymes_nozero_rows<-above1000_tpm_cazymes_nozero %>% mutate(sumVar = rowSums(.[1:298]))
rownames(above1000_tpm_cazymes_nozero_rows)<-rownames(above1000_tpm_cazymes_nozero)
above1000_tpm_cazymes_nozero_rows<-above1000_tpm_cazymes_nozero_rows%>%select(sumVar)

above1000_tpm_cazymes_nozero_rows$X<-rownames(above1000_tpm_cazymes_nozero_rows)


sampleinfo<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv")
above1000_tpm_cazymes_nozero_rows<-merge(above1000_tpm_cazymes_nozero_rows,sampleinfo,by.x="X",by.y="termite_tree_labels")

setDT(above1000_tpm_cazymes_nozero_rows)
relativecazymes_diet<-above1000_tpm_cazymes_nozero_rows[, .(sumVar = sum(sumVar)), by = .(diet_type_number4)]


head(relativecazymes_diet,5)
 

##NOTE- both tables, binary counts of CAZYmes in 4 diets and relative abundance of CAZYmes in 4 diets show that diet1 and diet2 are have more CAZYmes than diet3 and diet5.
#---------------------------------------------------------------------------------

```



## Which CAZymes are more abundant in Fungal feeders and Lower termites (TPM values)-
```{r}

##relative abundance of GH for enzymes significantly different in Macrotermitinae as compared to other wood-feeding Termitidae-
phyloanova<-read.csv("Phylogenetic_anova_withdiet_153CAZYmes_122samples_feb2021.csv",row.names=1)
phyloanova_macro<-phyloanova%>%filter(X2.5 <=0.05) #get significantly different GHs in diet5 vs diet2


#TPM abundance of each CAZYme per diet-
above1000_tpm_cazymes_nozero$X<-rownames(above1000_tpm_cazymes_nozero)
sampleinfo<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv")
above1000_tpm_cazymes_nozero_diets<-merge(above1000_tpm_cazymes_nozero,sampleinfo,by.x="X",by.y="termite_tree_labels")

cols<-colnames(above1000_tpm_cazymes_nozero)
cols<-cols[-299]

setDT(above1000_tpm_cazymes_nozero_diets)
above1000_tpm_cazymes_nozero_diets2<-above1000_tpm_cazymes_nozero_diets[, lapply(.SD, sum, na.rm=TRUE), by=diet_type_number4, .SDcols=cols ]
above1000_tpm_cazymes_nozero_diets2<-as.data.frame(above1000_tpm_cazymes_nozero_diets2)
above1000_tpm_cazymes_nozero_diets2$diet_type_number4<-paste("diet",above1000_tpm_cazymes_nozero_diets2$diet_type_number4,sep="_")

#-------------------------------------------------------------------------
#abundance in Macrotermitinae-
rownames(above1000_tpm_cazymes_nozero_diets2)<-above1000_tpm_cazymes_nozero_diets2$diet_type_number4
above1000_tpm_cazymes_nozero_diets2$diet_type_number4<-NULL
above1000_tpm_cazymes_nozero_diets2_t<-t(above1000_tpm_cazymes_nozero_diets2)%>%as.data.frame()
above1000_tpm_cazymes_nozero_diets2_t_selected<-above1000_tpm_cazymes_nozero_diets2_t%>%filter(diet_5>diet_2) #all cazymes abundant in Macrotermitinae as compared to wood-feeding Termitinae.

above1000_tpm_cazymes_nozero_diets2_t_selected$genename<-rownames(above1000_tpm_cazymes_nozero_diets2_t_selected)
above1000_tpm_cazymes_nozero_diets2_t_selected$genename<-gsub(".hmm","",above1000_tpm_cazymes_nozero_diets2_t_selected$genename)

macro_abundant_phylanova<-merge(above1000_tpm_cazymes_nozero_diets2_t_selected,phyloanova_macro,by="genename")
write.csv(macro_abundant_phylanova,file="cazymes_abundantinMacrotermitinae_vs_Termitinae_feb2021.csv")

##NOTE-Out of 63 CAZYmes significantly abundant in Macrotermitinae vs Wood-feeding higher termites, 33 are more abundant in Macrotermitinae samples.

#reduced in Macrotermitinae-
above1000_tpm_cazymes_nozero_diets2_t_selected<-above1000_tpm_cazymes_nozero_diets2_t%>%filter(diet_5<diet_2) #all cazymes reduced in Macrotermitinae as compared to wood-feeding Termitinae.

above1000_tpm_cazymes_nozero_diets2_t_selected$genename<-rownames(above1000_tpm_cazymes_nozero_diets2_t_selected)
above1000_tpm_cazymes_nozero_diets2_t_selected$genename<-gsub(".hmm","",above1000_tpm_cazymes_nozero_diets2_t_selected$genename)

macro_reduced_phylanova<-merge(above1000_tpm_cazymes_nozero_diets2_t_selected,phyloanova_macro,by="genename")
write.csv(macro_reduced_phylanova,file="cazymes_reducedinMacrotermitinae_vs_Termitinae_feb2021.csv")

##NOTE-Out of 63 CAZYmes significantly abundant in Macrotermitinae vs Wood-feeding higher termites, 30 are reduced in Macrotermitinae samples.

#----------------------------------------------------------------------------
##Cazymes abundant in lower termites-
above1000_tpm_cazymes_nozero_diets2_t_selected #for abundance of individual genes.
phyloanova #for significance with diet

#phyloanova_lower<-phyloanova%>%filter(X1.2 <=0.05 | X1.3 <=0.05 | X1.5 <=0.05 ) #get significantly different GHs
phyloanova_lower<-phyloanova%>%filter(X1.2 <=0.05) #significantly different GHs in lower termites vs wood-feeding higher termites

above1000_tpm_cazymes_nozero_diets2_t_selected<-above1000_tpm_cazymes_nozero_diets2_t%>%filter(diet_1>diet_2) #all cazymes abundant in lower termites as compared to wood-feeding higher.

above1000_tpm_cazymes_nozero_diets2_t_selected$genename<-rownames(above1000_tpm_cazymes_nozero_diets2_t_selected)
above1000_tpm_cazymes_nozero_diets2_t_selected$genename<-gsub(".hmm","",above1000_tpm_cazymes_nozero_diets2_t_selected$genename)

lower_abundant_phylanova<-merge(above1000_tpm_cazymes_nozero_diets2_t_selected,phyloanova_lower,by="genename")
write.csv(lower_abundant_phylanova,file="cazymes_abundantLowertermitesvsWoodfeedinghigher_feb2021.csv")


#NOTE- Out of 91 CAZYmes signficant in phyl.anova() for lower termites, 36 are abundant in lower termites vs wood-feeding higher termites.

##CAZYmes reduced in lower termites-
above1000_tpm_cazymes_nozero_diets2_t_selected<-above1000_tpm_cazymes_nozero_diets2_t%>%filter(diet_1<diet_2) #all cazymes reduced in lower termites as wood-feeding higher termites.

above1000_tpm_cazymes_nozero_diets2_t_selected$genename<-rownames(above1000_tpm_cazymes_nozero_diets2_t_selected)
above1000_tpm_cazymes_nozero_diets2_t_selected$genename<-gsub(".hmm","",above1000_tpm_cazymes_nozero_diets2_t_selected$genename)

lower_reduced_phylanova<-merge(above1000_tpm_cazymes_nozero_diets2_t_selected,phyloanova_lower,by="genename")
write.csv(lower_reduced_phylanova,file="cazymes_reducedLowertermitesvsWoodfeedinghigher_feb2021.csv")

#NOTE-Out of 91 CAZYmes significant in phyl.anova() for lower termites, 55 are reduced in lower termites vs wood-feeding higher termites.

above1000_tpm_cazymes_nozero_diets2%>%select(GH9.hmm)
above1000_tpm_cazymes_nozero_diets2%>%select(GH13_8.hmm)
above1000_tpm_cazymes_nozero_diets2%>%select(GH92.hmm)
above1000_tpm_cazymes_nozero_diets2%>%select(GH133.hmm)

```


##Relative abundance in clr transformed data -TableS7
```{r}

##abundance per diet in clr transformed data-
cazymes_clr_10perc_3above<-read.csv("cazymes_153columns_122samples_clr_feb2021.csv")
sampleinfo<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv")
cazymes_clr_10perc_3above2<-merge(cazymes_clr_10perc_3above,sampleinfo,by.x="X",by.y="termite_tree_labels")
cols<-colnames(cazymes_clr_10perc_3above)
cols<-cols[-1]
setDT(cazymes_clr_10perc_3above2)
cazymes_clr_10perc_3above2_diet<-cazymes_clr_10perc_3above2[, lapply(.SD, sum, na.rm=TRUE), by=diet_type_number4, .SDcols=cols ] #for all the cazymes


write.csv(cazymes_clr_10perc_3above2_diet,file="cazymes_clr_perdiet_fortableS7_feb2021.csv")

#list of CAZYmes discussed in lower termites section in paper-
cazymes_clr_10perc_3above2_diet$diet_type_number4<-gsub("^","diet_",cazymes_clr_10perc_3above2_diet$diet_type_number4)
cazymes_clr_10perc_3above2_diet<-as.data.frame(cazymes_clr_10perc_3above2_diet)
rownames(cazymes_clr_10perc_3above2_diet)<-cazymes_clr_10perc_3above2_diet$diet_type_number4

cazymes_clr_10perc_3above2_diet%>%select(GH106.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH123.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH28.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH5_4.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH5_10.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH9.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH45.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH11.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH53.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH5_4.hmm)

#list of CAZYmes discussed in fungus-cultivating section in paper-
cazymes_clr_10perc_3above2_diet%>%select(GH44.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH45.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH8.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH10.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH5_10.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH11.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH133.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH109.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH97.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH78.hmm)
cazymes_clr_10perc_3above2_diet%>%select(GH28.hmm)


```
