---
title: "metabolic_genes"
author: "Jigyasa_Arora"
date: "2/28/2021"
---

```{r}

library(dplyr)
library(tidyr)
library(ape)
library(data.table)
```

##Extracting out metabolic genes from contigs+summing by subunit+using Hydrogenases classified by Hyddb-
```{r}  
 
functions<-read.csv("functions/all-metagenome-annotation-taxonomy-1000bps.txt")
functions$domain<-gsub("_p__.*$","",functions$taxa)

#get bacterial contigs out-
functions_bacteria<-functions%>%filter(domain=="d__Bacteria")

#sum by annotation + tpm-
setDT(functions_bacteria)
functions_bacteria_genes<-functions_bacteria[, .(TPM = sum(TPM)), by = .(sample_names,annotation)]

#spread the data-
library(tidyr)
functions_bacteria_genes_spread<-as.data.frame(spread(functions_bacteria_genes,annotation,TPM))
rownames(functions_bacteria_genes_spread)<-functions_bacteria_genes_spread$sample_names


#get 129 samples out-
#termite time tree
newtree_129samples<-read.tree("1000bps_taxonomy_129taxa_tree.nwk")
dnew2<-data.frame(label=newtree_129samples$tip.label)

dnew2$label<-as.character(dnew2$label)
dnew2$runnumber<-unlist(lapply(strsplit(dnew2$label,split="-"),"[",1))
dnew2$runnumber<-gsub("_","-",dnew2$runnumber)

functions_bacteria_genes_spread2<-merge(functions_bacteria_genes_spread,dnew2,by.x="sample_names",by.y="runnumber",all.y=TRUE)
rownames(functions_bacteria_genes_spread2)<-functions_bacteria_genes_spread2$label

#order the rownames by termite tree to get samples with no data at this step!
functions_bacteria_genes_spread2<-functions_bacteria_genes_spread2[match(newtree_129samples$tip.label,rownames(functions_bacteria_genes_spread2)), ]

write.csv(functions_bacteria_genes_spread2,file="bacterial_functionalgenes_all_march2021.csv")
#--------------------------------------------------------------------------------------
##summing by subunits-
main_funcs<-read.csv("functions/allbacterial_koids_phylanova_jan2021.csv",header=FALSE)

#1.
acscols<-c("K00198","K00194","K00197") #"K14138", "K15023" not found.
functions_bacteria_genes_spread2$acs<-functions_bacteria_genes_spread2%>%select("K00198","K00194","K00197")%>%rowSums(na.rm=TRUE)
functions_bacteria_genes_spread2<-functions_bacteria_genes_spread2%>%select(-c("K00198","K00194","K00197"))

#2.
apr<-c("K00394","K00395")
functions_bacteria_genes_spread2$apr<-functions_bacteria_genes_spread2%>%select("K00394","K00395")%>%rowSums(na.rm=TRUE)
functions_bacteria_genes_spread2<-functions_bacteria_genes_spread2%>%select(-c("K00394","K00395"))

#3.
nif<-c("K02586","K02591","K02588")
functions_bacteria_genes_spread2$nif<-functions_bacteria_genes_spread2%>%select("K02586","K02591","K02588")%>%rowSums(na.rm=TRUE)
functions_bacteria_genes_spread2<-functions_bacteria_genes_spread2%>%select(-c("K02586","K02591","K02588"))

#4. 
nar<-c("K00370","K00371","K00374")
functions_bacteria_genes_spread2$nar<-functions_bacteria_genes_spread2%>%select("K00370","K00371","K00374")%>%rowSums(na.rm=TRUE)
functions_bacteria_genes_spread2<-functions_bacteria_genes_spread2%>%select(-c("K00370","K00371","K00374"))

#5.
nap<-c("K02567","K02568")
functions_bacteria_genes_spread2$nap<-functions_bacteria_genes_spread2%>%select("K02567","K02568")%>%rowSums(na.rm=TRUE)
functions_bacteria_genes_spread2<-functions_bacteria_genes_spread2%>%select(-c("K02567","K02568"))

#6.
nir<-c("K00362","K00363")
functions_bacteria_genes_spread2$nir<-functions_bacteria_genes_spread2%>%select("K00362","K00363")%>%rowSums(na.rm=TRUE)
functions_bacteria_genes_spread2<-functions_bacteria_genes_spread2%>%select(-c("K00362","K00363"))

#7.
nrf<-c("K03385") #only nrfA present.K15876 absent
functions_bacteria_genes_spread2$nrf<-functions_bacteria_genes_spread2%>%select("K03385")%>%rowSums(na.rm=TRUE)
functions_bacteria_genes_spread2<-functions_bacteria_genes_spread2%>%select(-c("K03385"))

#8.
ure<-c("K01428","K01429","K01430")
functions_bacteria_genes_spread2$ure<-functions_bacteria_genes_spread2%>%select("K01428","K01429","K01430")%>%rowSums(na.rm=TRUE)
functions_bacteria_genes_spread2<-functions_bacteria_genes_spread2%>%select(-c("K01428","K01429","K01430"))

#9.
glt<-c("K00265","K00266")
functions_bacteria_genes_spread2$glt<-functions_bacteria_genes_spread2%>%select("K00265","K00266")%>%rowSums(na.rm=TRUE)
functions_bacteria_genes_spread2<-functions_bacteria_genes_spread2%>%select(-c("K00265","K00266"))

#10.
arg<-c("K00611","K01940","K01755")
functions_bacteria_genes_spread2$arg<-functions_bacteria_genes_spread2%>%select("K00611","K01940","K01755")%>%rowSums(na.rm=TRUE)
functions_bacteria_genes_spread2<-functions_bacteria_genes_spread2%>%select(-c("K00611","K01940","K01755"))

#11.
nife_group1<-c("[NiFe] Group 1a","[NiFe] Group 1b","[NiFe] Group 1c","[NiFe] Group 1d","[NiFe] Group 1f","[NiFe] Group 1i")
functions_bacteria_genes_spread2$Group1<-functions_bacteria_genes_spread2%>%select("[NiFe] Group 1a","[NiFe] Group 1b","[NiFe] Group 1c","[NiFe] Group 1d","[NiFe] Group 1f","[NiFe] Group 1i")%>%rowSums(na.rm=TRUE)
functions_bacteria_genes_spread2<-functions_bacteria_genes_spread2%>%select(-c("[NiFe] Group 1a","[NiFe] Group 1b","[NiFe] Group 1c","[NiFe] Group 1d","[NiFe] Group 1f","[NiFe] Group 1i"))


#12. 
colnames(functions_bacteria_genes_spread2)[which(names(functions_bacteria_genes_spread2) == "[NiFe] Group 2b")] <- "Group2"

#13.
nife_group3<-c("[NiFe] Group 3a","[NiFe] Group 3b","[NiFe] Group 3d")
functions_bacteria_genes_spread2$Group3<-functions_bacteria_genes_spread2%>%select("[NiFe] Group 3a","[NiFe] Group 3b","[NiFe] Group 3d")%>%rowSums(na.rm=TRUE)
functions_bacteria_genes_spread2<-functions_bacteria_genes_spread2%>%select(-c("[NiFe] Group 3a","[NiFe] Group 3b","[NiFe] Group 3d"))

#14. 
nife_group4<-c("[NiFe] Group 4a","[NiFe] Group 4c","[NiFe] Group 4e","[NiFe] Group 4g")
functions_bacteria_genes_spread2$Group4<-functions_bacteria_genes_spread2%>%select("[NiFe] Group 4a","[NiFe] Group 4c","[NiFe] Group 4e","[NiFe] Group 4g")%>%rowSums(na.rm=TRUE)
functions_bacteria_genes_spread2<-functions_bacteria_genes_spread2%>%select(-c("[NiFe] Group 4a","[NiFe] Group 4c","[NiFe] Group 4e","[NiFe] Group 4g"))


#15.
fefe_groupC<-c("Group C1","Group C2","Group C3")
functions_bacteria_genes_spread2$GroupC<-functions_bacteria_genes_spread2%>%select("Group C1","Group C2","Group C3")%>%rowSums(na.rm=TRUE)
functions_bacteria_genes_spread2<-functions_bacteria_genes_spread2%>%select(-c("Group C1","Group C2","Group C3"))

#16.
colnames(functions_bacteria_genes_spread2)[which(names(functions_bacteria_genes_spread2) == "Group A")] <- "GroupA"
colnames(functions_bacteria_genes_spread2)[which(names(functions_bacteria_genes_spread2) == "Group B")] <- "GroupB"

write.csv(functions_bacteria_genes_spread2,file="allfunctions_rawTPM_129samples_genesnosubunit.csv")

```


##Filters and clr transformation of data-
```{r} 
##Transformations-

##clr transform-
clr_transform<-function(filename_spread2){ 
library(propr) 
filename_spread2[is.na(filename_spread2)] <-0 #convert NA to zero
#rownames(filename_spread2)<-filename_spread2$label
filename_spread2$labels<-NULL
filename_spread2$sample_names<-NULL
#add psuedo counts to account for missing and not-present data-
filename_spread2<-filename_spread2+0.65

filename_propr<-propr(filename_spread2,metric="rho",ivar="clr")

#creating a df with all the transformed pathways and metadata-
filename_propr_df<-filename_propr@logratio

return(filename_propr_df)
}

functions_clr<-clr_transform(functions_bacteria_genes_spread2)

#finding non-numeric columns in a large data-frame-
functions_bacteria_genes_spread2[is.na(functions_bacteria_genes_spread2)] <-0
functions_bacteria_genes_spread2$sample_names<-NULL
functions_bacteria_genes_spread2$label<-NULL
original_cols<-colnames(functions_bacteria_genes_spread2)%>%as.data.frame()
colnames(original_cols)<-c("columns")

test<-functions_bacteria_genes_spread2%>% select_if(is.numeric)
numeric_columns<-colnames(test)%>%as.data.frame()
colnames(numeric_columns)<-c("columns")

removedcols<-anti_join(original_cols,numeric_columns,by="columns")

#-------------------------------------------------------------------------------------
#select the keggs from metabolic pathways out-
main_funcs<-read.csv("functions/allbacterial_koids_metabolicgenesmetabolicgenes_nosubunits_jan2021.csv",header=FALSE) #genes with subunits have correct names in this file

mainfuncs_functions_clr<-functions_clr%>%dplyr::select_if(names(.) %in% main_funcs$V1) #select the keggs out
write.csv(mainfuncs_functions_clr,file="metabolic_pathways_clr_march2021.csv")

mainfuncs_functions_tpm<-functions_bacteria_genes_spread2%>%dplyr::select_if(names(.) %in% main_funcs$V1)
write.csv(mainfuncs_functions_tpm,file="metabolic_pathways_rawtpm_march2021.csv")

#---------------------------------------------------------------------------------------
##select annotations present in more than 7.5% of the samples-(9/129)
above7.5perc_function<-function(filename){
  filename$label<-NULL
  filename$samplename<-NULL
 
  filename2<-filename #for filtering
  filename2[filename2 > 0] <- 1
  filename2[filename2 <= 0] <-0
  filename_10perc<-filename2[colSums(filename2) >=9.6]
  columns<-colnames(filename_10perc)

  filename3<-filename%>%dplyr::select(columns)
  return(filename3)
}

functions_above7.5<-above7.5perc_function(mainfuncs_functions_clr)

morethan3taxa_functions<-function(filename){

filename$label<-NULL
filename$samplename<-NULL
filename2<-filename
filename2[filename2 > 0] <- 1
filename2[filename2 <= 0] <-0
filename_threeabove<-filename2[rowSums(filename2[,])>=3, ]
rows<-rownames(filename_threeabove)

filename3<-subset(filename, rownames(filename) %in% rows)

return(filename3)
}

functions_above7.5_3above<-morethan3taxa_functions(functions_above7.5)
write.csv(functions_above7.5_3above,file="metabolic_pathways_clr_126samples_genesnosubunit.csv")

#----------------------------------------------------------------------------------------
#order the columns by metabolic pathways-
main_funcs$V1[!(main_funcs$V1) %in% colnames(functions_above7.5_3above)] #find keggs not present

main_funcs2<-main_funcs%>%filter(!V1 %in% c("K13990","K00122","K11180","K11181","Group2")) #keep genes in metagenome contigs 
genenames<-as.vector(main_funcs2$V1)
functions_above7.5_3above_ordered<-functions_above7.5_3above[,genenames] #order the metagenome contigs data


#order by termite tree-
tree<-read.tree("1000bps_taxonomy_129taxa_tree.nwk")
newtree<-keep.tip(tree,as.vector(rownames(functions_above7.5_3above_ordered)))

functions_above7.5_3above_ordered<-functions_above7.5_3above_ordered[match(newtree$tip.label,rownames(functions_above7.5_3above_ordered)), ]

write.csv(functions_above7.5_3above_ordered,file="metabolic_pathways_clr_126samples_genesnosubunit.csv")
write.tree(newtree,file="1000bps_taxonomy_126taxa_tree.nwk")




```

##Moran's I and Phyl.anova stats-
```{r}
#Moran's I-
functions<-read.csv("metabolic_pathways_clr_126samples_genesnosubunit.csv",row.names = 1) #clr transformed + summed by subunit
tree<-read.tree("1000bps_taxonomy_126taxa_tree.nwk")

functions<-functions[match(tree$tip.label,rownames(functions)), ] #match the tree



#run phylo.signal-<get signficiant p.values for phylogenetic correlation for each taxa>
library(adephylo)
library(ape)
library(phylobase)
library(phylosignal)

#clr transformed data with 0 values-<all funcs>
p4d <- phylo4d(tree,functions)
p2<-phyloSignal(p4d = p4d, method = "I")

#adjusted p-value using Benjamini & Hochberg method ("BH") or fdr method-
mod2 = as.data.frame(p2$pvalue)
mod2$padjusted_lambda<-p.adjust(mod2$I, method='fdr')
mod2$functions<-rownames(mod2)
mod2<-mod2%>%dplyr::mutate(significance_padjusted=ifelse(padjusted_lambda<0.05,"*",""))
mod2<-mod2%>%mutate(significance_I=ifelse(I<0.05,"*",""))
main_funcs<-read.csv("functions/allbacterial_koids_metabolicgenes_nosubunits_jan2021.csv",header=FALSE)
main_funcs$V1<-gsub(" ",".",main_funcs$V1) #match the column with colnames of file.
mod2<-merge(mod2,main_funcs,by.x="functions",by.y="V1")

write.csv(mod2,file="phylogenetic_autocorrelation_metabolicgenes_nosubunit_126samples_march2021.csv")
#-------------------------------------------------------------------------------------------------------------
##phyl.anova-
library(geomorph)
functions<-functions+100 #add a constant to every value

sampleinfo<-read.csv("taxonomy/sample_names_ids_withotherfactors_tomedit.2.csv",row.names=1)
rownames(sampleinfo)<-sampleinfo$termite_tree_labels
sampleinfo<-sampleinfo[match(tree$tip.label,rownames(sampleinfo)), ] #match the tree

phyl.anova.func2_list<-function(filename){
filename<-as.data.frame(t(filename)) #convert row to column
filename$label<-rownames(filename)
geo_functions<-merge(filename,sampleinfo,by.x="label",by.y="termite_tree_labels")
geo_functions$diet_type_number4<-as.factor(geo_functions$diet_type_number4)
geo_functions[,2]<-as.numeric(geo_functions[,2])

gdf <- geomorph.data.frame(abundance = geo_functions[,2], diet = as.factor(geo_functions$diet_type_number4), phy = tree) #create a dataframe that contains the data and phylogeny. [according to google forum]

#group_name <- names(filename)[2] #get the correct column name. column 2 is the cazyme.

geomorph.lm<-procD.lm(abundance ~ diet, iter=99999,RRPP=TRUE,effect.type = "F", SS.type = "II" ,data=gdf,print.progress = FALSE) #SS.type="II" is used if the groups (i.e. diet/lineage) are of uneven sizes.
mod<-anova(geomorph.lm) #to examine if there is a phylogenetic correlation between "y" and "x"
mod2<-as.data.frame(mod$table)
mod2<-mod2[1,7]
# d is distance between vector lengths is the difference in the amount of shape change per unit of size. If d is 0.05 i.e., | 0.20 - 0.15 |.  Why do this?  Two vectors can be correlated in direction - the way shape changes with size - but one group can exhibit more shape change per unit size than the other
#D tells us the effect size of the standard deviation between two groups. Positive and above 0.2-0.3 means larger positive difference, while negative and below -0.2 to -0.3 means larger negative difference.

pw<-pairwise(geomorph.lm,groups = as.factor(gdf$diet))
pw_summary<-summary(pw,confidence=0.95,show.vectors = TRUE)
pw_summary2<-as.data.frame(pw_summary$summary.table)

pw_summary2$padjust<-p.adjust(pw_summary2$`Pr > d`,method="fdr") #fdr correction of p.value
pw_summary2$diets<-rownames(pw_summary2)

#convert the output to correct formats-
names<-colnames(geo_functions[2])
pw_summary2$genename<-rep(names,times=nrow(pw_summary2))
library(tidyr)
pw_summary3<-pw_summary2%>%dplyr::select("padjust","diets","genename")
pw_spread<-pw_summary3%>%spread(diets,padjust)
pw_spread2<-cbind(pw_spread,mod2)

return(pw_spread2)
}

functions_t<-as.data.frame(t(functions))
xy.list <-  split(functions_t, seq(nrow(functions_t))) #convert df to a list

kegg_geomorph<-lapply(xy.list,phyl.anova.func2_list) #apply on all columns

kegg_geomorph_df<-do.call(rbind.data.frame,kegg_geomorph)


#find p.adjust for anova-
kegg_geomorph_df$padjust<-p.adjust(kegg_geomorph_df$mod2)

kegg_geomorph_df<-kegg_geomorph_df%>%mutate(significant_anova=ifelse(padjust <=0.05,"*",""))

#join with genenames-
main_funcs<-read.csv("functions/allbacterial_koids_metabolicgenes_nosubunits_jan2021.csv",header=FALSE)
main_funcs$V1<-gsub(" ",".",main_funcs$V1) #match the column with colnames of file.
kegg_geomorph_df2<-merge(kegg_geomorph_df,main_funcs,by.x="genename",by.y="V1")

write.csv(kegg_geomorph_df2,file="Phylogenetic_anova_withdiet_metabolicpathwaygenes_nosubunit_126samples_march2021.csv")


#--------------------------------------------------------------------------------
#Supplementary Table S11-

phylanova<-read.csv("Phylogenetic_anova_withdiet_metabolicpathwaygenes_nosubunit_126samples_march2021.csv")
phylanova<-phylanova%>%mutate(new1.02=ifelse(X1.2<0.05 & X1.2>0.01,"*",ifelse(X1.2<0.01 & X1.2>0.001,"**",ifelse(X1.2<0.001,"***","not-significant"))))
phylanova<-phylanova%>%mutate(new1.03=ifelse(X1.3<0.05 & X1.3>0.01,"*",ifelse(X1.3<0.01 & X1.3>0.001,"**",ifelse(X1.3<0.001,"***","not-significant"))))
phylanova<-phylanova%>%mutate(new1.05=ifelse(X1.5<0.05 & X1.5>0.01,"*",ifelse(X1.5<0.01 & X1.5>0.001,"**",ifelse(X1.5<0.001,"***","not-significant"))))
phylanova<-phylanova%>%mutate(new2.03=ifelse(X2.3<0.05 & X2.3>0.01,"*",ifelse(X2.3<0.01 & X2.3>0.001,"**",ifelse(X2.3<0.001,"***","not-significant"))))
phylanova<-phylanova%>%mutate(new2.05=ifelse(X2.5<0.05 & X2.5>0.01,"*",ifelse(X2.5<0.01 & X2.5>0.001,"**",ifelse(X2.5<0.001,"***","not-significant"))))
phylanova<-phylanova%>%mutate(new3.05=ifelse(X3.5<0.05 & X3.5>0.01,"*",ifelse(X3.5<0.01 & X3.5>0.001,"**",ifelse(X3.5<0.001,"***","not-significant"))))

write.csv(phylanova,file="Phylogenetic_anova_withdiet_metabolicpathwaygenes_nosubunit_126samples_march2021.csv")
```

##Basic stats on metabolic genes-
```{r} 


mainfuncs_tpm<-read.csv("metabolic_pathways_rawtpm_march2021.csv")
rownames(mainfuncs_tpm)<-mainfuncs_tpm$X

##relative abundance of each gene across diets-<raw TPM>

sampleinfo<-read.csv("taxonomy/sample_names_ids_withotherfactors_tomedit.2.csv")
sampleinfo<-sampleinfo%>%select(termite_tree_labels,diet_type_number4)

mainfuncs_tpm_diet<-merge(mainfuncs_tpm,sampleinfo,by.x="X",by.y="termite_tree_labels")

#get 126 samples out-as basic stats should match the phylo stats-
tree<-read.tree("1000bps_taxonomy_126taxa_tree.nwk")
dnew2<-data.frame(label=tree$tip.label)

dnew2$label<-as.character(dnew2$label)
dnew2$runnumber<-unlist(lapply(strsplit(dnew2$label,split="-"),"[",1))
dnew2$runnumber<-gsub("_","-",dnew2$runnumber)

mainfuncs_tpm_diet<-merge(mainfuncs_tpm_diet,dnew2,by.x="X",by.y="label",all.y=TRUE)
rownames(mainfuncs_tpm_diet)<-mainfuncs_tpm_diet$X


#sum by diet-
cols<-colnames(mainfuncs_tpm_diet)
cols<-cols[!cols %in% c("X","diet_type_number4","runnumber")]

setDT(mainfuncs_tpm_diet)
mainfuncs_tpm_diet2<-mainfuncs_tpm_diet[, lapply(.SD, sum, na.rm=TRUE), by=diet_type_number4, .SDcols=cols ] #for all the metabolic functions
mainfuncs_tpm_diet2$diet_type_number4<-gsub("^","diet_",mainfuncs_tpm_diet2$diet_type_number4)
mainfuncs_tpm_diet2<-as.data.frame(mainfuncs_tpm_diet2)
rownames(mainfuncs_tpm_diet2)<-mainfuncs_tpm_diet2$diet_type_number4


#checking the relative abundance of reductive acetogenesis genes per diet-
mainfuncs_tpm_diet2%>%select("K01938","K01491","K00297","acs","K00625","K00925")


write.csv(mainfuncs_tpm_diet2,file="functions-metabolicgenes-bydiet-march2021.csv")

#presence of metabolic genes per termite lineage-
mainfuncs_tpm<-read.csv("metabolic_pathways_rawtpm_march2021.csv",row.names=1)

sampleinfo<-read.csv("taxonomy/sample_names_ids_withotherfactors_tomedit.2.csv")
sampleinfo<-sampleinfo%>%select(termite_tree_labels,diet_type_number4,lineage,lineage2)

reductive<-mainfuncs_tpm%>%select("K01938","K01491","K00297","K00122","acs","K00625","K00925")%>%filter_all(any_vars(. != 0)) #128 samples
sulfate<-mainfuncs_tpm%>%select("apr","K00958")%>%filter_all(any_vars(. != 0)) #65 samples
sulfate$X<-rownames(sulfate)
sulfate<-merge(sulfate,sampleinfo,by.x="X",by.y="termite_tree_labels")


#---------------------------------------------------------------------------------
#what is the relative abundance of hydrogenases in the metagenomes?
hydrogenases<-read.csv("metabolic_pathways_rawtpm_march2021.csv")
hydrogenases[is.na(hydrogenases)] <-0

#get 126 samples out-as basic stats should match the phylo stats-
library(ape)
tree<-read.tree("1000bps_taxonomy_126taxa_tree.nwk")
dnew2<-data.frame(label=tree$tip.label)

dnew2$label<-as.character(dnew2$label)
dnew2$runnumber<-unlist(lapply(strsplit(dnew2$label,split="-"),"[",1))
dnew2$runnumber<-gsub("_","-",dnew2$runnumber)

hydrogenases<-merge(hydrogenases,dnew2,by.x="X",by.y="label",all.y=TRUE)
rownames(hydrogenases)<-hydrogenases$X


fefe_overallsum<-sum(hydrogenases$GroupA)+sum(hydrogenases$GroupB)+sum(hydrogenases$GroupC)
nife_overallsum<-sum(hydrogenases$Group1)+sum(hydrogenases$Group2)
hydrogenases_overallsum<-fefe_overallsum+nife_overallsum

print(paste0("Overall abundance of FeFe hydrogenases is ", fefe_overallsum/(nife_overallsum) , " fold more than NiFe hydrogenases"))


print(paste0("Overall abundance of GroupA FeFe hydrogenases is ", sum(hydrogenases$GroupA)/fefe_overallsum*100))
print(paste0("Overall abundance of GroupB FeFe hydrogenases is ", sum(hydrogenases$GroupB)/fefe_overallsum*100))
print(paste0("Overall abundance of GroupC FeFe hydrogenases is ", sum(hydrogenases$GroupC)/fefe_overallsum*100))

print(paste0("Overall abundance of Group1 NiFe hydrogenases is ", sum(hydrogenases$Group1)/nife_overallsum*100))
print(paste0("Overall abundance of Group2 NiFe hydrogenases is ", sum(hydrogenases$Group2)/nife_overallsum*100))


#across diets-
#checking the relative abundance of genes per diet-<mainfuncs_tpm_diet2 is based on 126 samples>
mainfuncs_tpm_diet2%>%select(GroupA,GroupC)
mainfuncs_tpm_diet2%>%select(Group1)

#distribution of hydrogenases across termite lineages-
##select hydrogenases, remove termite samples with zero values
hydrogenases2<-hydrogenases%>%select(GroupA,GroupB,GroupC,Group1,Group2)%>%filter_all(any_vars(. != 0)) 
hydrogenases2$X<-rownames(hydrogenases2)

sampleinfo<-read.csv("taxonomy/sample_names_ids_withotherfactors_tomedit.2.csv")
sampleinfo<-sampleinfo%>%select(termite_tree_labels,diet_type_number4,lineage,lineage2)

hydrogenases2<-merge(hydrogenases2,sampleinfo,by.x="X",by.y="termite_tree_labels")
hydrogenases2%>%select(lineage)%>%group_by(lineage)%>%count()
#--------------------------------------------------------------------------------------
#NiFe sub-classification-
functions_bacteria_genes_spread<-read.csv("bacterial_functionalgenes_all_march2021.csv")
#functions_bacteria_genes_spread<-as.data.frame(spread(functions_bacteria_genes,annotation,TPM))
#rownames(functions_bacteria_genes_spread)<-functions_bacteria_genes_spread$sample_names
functions_bacteria_genes_spread[is.na(functions_bacteria_genes_spread)] <-0

#get 126 samples out-
tree<-read.tree("1000bps_taxonomy_126taxa_tree.nwk")
dnew2<-data.frame(label=tree$tip.label)

dnew2$label<-as.character(dnew2$label)
dnew2$runnumber<-unlist(lapply(strsplit(dnew2$label,split="-"),"[",1))
dnew2$runnumber<-gsub("_","-",dnew2$runnumber)

functions_bacteria_genes_spread2<-merge(functions_bacteria_genes_spread,dnew2,by.x="sample_names",by.y="runnumber",all.y=TRUE)
rownames(functions_bacteria_genes_spread2)<-functions_bacteria_genes_spread2$label

functions_bacteria_genes_spread2[is.na(functions_bacteria_genes_spread2)] <-0
rownames(functions_bacteria_genes_spread2)<-functions_bacteria_genes_spread2$X

#get the NiFe groups out-
columns<-colnames(functions_bacteria_genes_spread2)%>%as.data.frame()
colnames(columns)<-c("cols")
nifecols<-columns%>%filter(str_detect(cols, "NiFe"))%>%as.vector()

#extract NiFe sub-classifications out-
nife_group1d<-functions_bacteria_genes_spread2%>%select("X.NiFe..Group.1d")
nife_allgroups<-functions_bacteria_genes_spread2%>%select(nifecols$cols)

allnife<-sum(nife_allgroups$X.NiFe..Group.1a)+sum(nife_allgroups$X.NiFe..Group.1b)+sum(nife_allgroups$X.NiFe..Group.1c)+sum(nife_allgroups$X.NiFe..Group.1d)+sum(nife_allgroups$X.NiFe..Group.1f)+sum(nife_allgroups$X.NiFe..Group.1i)+sum(nife_allgroups$X.NiFe..Group.2b)+sum(nife_allgroups$X.NiFe..Group.3a)+sum(nife_allgroups$X.NiFe..Group.3b)+sum(nife_allgroups$X.NiFe..Group.3d)+sum(nife_allgroups$X.NiFe..Group.4a)+sum(nife_allgroups$X.NiFe..Group.4c)+sum(nife_allgroups$X.NiFe..Group.4e)+sum(nife_allgroups$X.NiFe..Group.4g)

allnifegroup1<-sum(nife_allgroups$X.NiFe..Group.1a)+sum(nife_allgroups$X.NiFe..Group.1b)+sum(nife_allgroups$X.NiFe..Group.1c)+sum(nife_allgroups$X.NiFe..Group.1d)+sum(nife_allgroups$X.NiFe..Group.1f)+sum(nife_allgroups$X.NiFe..Group.1i)

print(paste0("Percent distribution of NiFe Group 1d in 126 metagenomes are ", sum(nife_group1d$X.NiFe..Group.1d)/allnife*100, "%"))
print(paste0("Percent distribution of NiFe Group 1d in 126 metagenomes are ", sum(nife_group1d$X.NiFe..Group.1d)/allnifegroup1*100, "%"))

#----------------------------------------------------------------------
##how many samples are sulfate reducing genes present?

mainfuncs_tpm<-read.csv("metabolic_pathways_rawtpm_march2021.csv")
rownames(mainfuncs_tpm)<-mainfuncs_tpm$X

mainfuncs_functions_sulfate<-mainfuncs_tpm%>%select(apr,K00958)

#convert everything to binary-
mainfuncs_functions_sulfate_binary<-mainfuncs_functions_sulfate
mainfuncs_functions_sulfate_binary[is.na(mainfuncs_functions_sulfate_binary)] <-0 #convert NA to zero
mainfuncs_functions_sulfate_binary[mainfuncs_functions_sulfate_binary > 0] <- 1 
mainfuncs_functions_sulfate_binary[mainfuncs_functions_sulfate_binary <= 0] <-0

mainfuncs_functions_sulfate_binary_rows<-mainfuncs_functions_sulfate_binary %>% mutate(sumVar = rowSums(.[1:2])) #get the sum of two genes
rownames(mainfuncs_functions_sulfate_binary_rows)<-rownames(mainfuncs_functions_sulfate)
mainfuncs_functions_sulfate_binary_rows<-mainfuncs_functions_sulfate_binary_rows%>%select(sumVar)
mainfuncs_functions_sulfate_binary_rows[mainfuncs_functions_sulfate_binary_rows > 0] <- 1  #presence/absence of apr and sat genes in all termite samples.

print(paste0("sulfate reducing genes aprAB and sat are found in ",sum(mainfuncs_functions_sulfate_binary_rows$sumVar)/nrow(mainfuncs_functions_sulfate_binary_rows)*100, "% of termite samples"))

#across diets-
#checking the relative abundance of genes per diet-<mainfuncs_tpm_diet2 is based on 126 samples>
mainfuncs_tpm_diet2%>%select(apr)

#----------------------------------------------------------------------------------------
##differences across diet for nirogen metabolism-
#across diets-
#checking the relative abundance of genes per diet-<mainfuncs_tpm_diet2 is based on 126 samples>
mainfuncs_tpm_diet2%>%select(ure)
mainfuncs_tpm_diet2%>%select(arg)
#----------------------------------------------------------------------------------------
##how many samples are nifHDK genes present?

mainfuncs_tpm<-read.csv("metabolic_pathways_rawtpm_march2021.csv",row.names = 1)


mainfuncs_tpm_nif<-mainfuncs_tpm%>%select(nif)
mainfuncs_tpm_nif$X<-rownames(mainfuncs_tpm_nif)

#what is the difference of nif genes across diet?
sampleinfo<-read.csv("taxonomy/sample_names_ids_withotherfactors_tomedit.2.csv")
sampleinfo<-sampleinfo%>%select(termite_tree_labels,diet_type_number4)

mainfuncs_tpm_nif_diet<-merge(mainfuncs_tpm_nif,sampleinfo,by.x="X",by.y="termite_tree_labels")
setDT(mainfuncs_tpm_nif_diet)
mainfuncs_tpm_nif_diet2<-mainfuncs_tpm_nif_diet[, .(nif = sum(nif)), by = .(diet_type_number4)]

lt_wf<-mainfuncs_tpm_nif_diet2[1,2]+mainfuncs_tpm_nif_diet2[2,2]
lt<-mainfuncs_tpm_nif_diet2[1,2]
wf<-mainfuncs_tpm_nif_diet2[2,2]
sf<-mainfuncs_tpm_nif_diet2[3,2]
fc<-mainfuncs_tpm_nif_diet2[4,2]

print(paste0("LT+WF are ", (lt_wf)/sf ," fold abundant than SF"))
print(paste0("LT+WF are ", (lt_wf)/fc ," fold abundant than FC"))
print(paste0("LT is ", (lt)/sf ," fold abundant than SF"))
print(paste0("LT is ", (lt)/fc ," fold abundant than FC"))
print(paste0("WF is ", (wf)/sf ," fold abundant than SF"))
print(paste0("WF is ", (wf)/fc ," fold abundant than FC"))

#-------------------------------------------------------------------------------


```




##methanogenesis in metagenomes genes- 
```{r} 
functions<-read.csv("functions/all-metagenome-annotation-taxonomy-1000bps.txt")
functions$domain<-gsub("_p__.*$","",functions$taxa)

#get archaeal contigs out-
functions_archaea<-functions%>%filter(domain=="d__Archaea")


#sum by annotation + tpm-
setDT(functions_archaea)
functions_archaea_genes<-functions_archaea[, .(TPM = sum(TPM)), by = .(sample_names,annotation)]

#spread the data-
library(tidyr)
functions_archaea_genes_spread<-as.data.frame(spread(functions_archaea_genes,annotation,TPM))
rownames(functions_archaea_genes_spread)<-functions_archaea_genes_spread$sample_names


#get 129 samples out-
#termite time tree
newtree_129samples<-read.tree("1000bps_taxonomy_129taxa_tree.nwk")
dnew2<-data.frame(label=newtree_129samples$tip.label)

dnew2$label<-as.character(dnew2$label)
dnew2$runnumber<-unlist(lapply(strsplit(dnew2$label,split="-"),"[",1))
dnew2$runnumber<-gsub("_","-",dnew2$runnumber)

functions_archaea_genes_spread2<-merge(functions_archaea_genes_spread,dnew2,by.x="sample_names",by.y="runnumber",all.y=TRUE)
rownames(functions_archaea_genes_spread2)<-functions_archaea_genes_spread2$label

#order the rownames by termite tree to get samples with no data at this step!
functions_archaea_genes_spread2<-functions_archaea_genes_spread2[match(newtree_129samples$tip.label,rownames(functions_archaea_genes_spread2)), ]


#extract mcr genes out-
mcr<-c("K00399","K00401","K00402")
functions_archaea_genes_spread2$mcr<-functions_archaea_genes_spread2%>%select("K00399","K00401","K00402")%>%rowSums(na.rm=TRUE)
functions_archaea_genes_spread2<-functions_archaea_genes_spread2%>%select(-c("K00399","K00401","K00402"))

write.csv(functions_archaea_genes_spread2,file="archaea_metabolicgenes-rawtpm_129samples.csv")
```

##filters and clr-transformation of methanogenesis data-
```{r}
functions<-read.csv("archaea_metabolicgenes-rawtpm_129samples.csv",row.names=1) 

#clr transform the data-
clr_transform<-function(filename_spread2){ 
library(propr) 
filename_spread2[is.na(filename_spread2)] <-0 #convert NA to zero
#rownames(filename_spread2)<-filename_spread2$label
filename_spread2$label<-NULL
filename_spread2$sample_names<-NULL
#add psuedo counts to account for missing and not-present data-
filename_spread2<-filename_spread2+0.65

filename_propr<-propr(filename_spread2,metric="rho",ivar="clr")

#creating a df with all the transformed pathways and metadata-
filename_propr_df<-filename_propr@logratio

return(filename_propr_df)
}

functions_clr<-clr_transform(functions)
#rownames(functions_clr)<-functions$label


write.csv(functions_clr,file="archaea_metabolicgenes-clr_129samples.csv")
#---------------------------------------------------------------------------------------------
#mcrABG methanogenesis genes to extract from clr transformed data-

##NOTE- Not performing any filtering ON mcrABG genes separetly as mcrABG genes are present in ~33 termite samples. So it already meets all the criteria for filtering.

functions_clr_mcr<-functions_clr%>%select(mcr)
write.csv(functions_clr_mcr,file="mcr_archaea_march2021.csv")


```

##stats on methanogenesis data-
```{r}
##tests-

tree<-read.tree("1000bps_taxonomy_129taxa_tree.nwk")

functions<-read.csv("mcr_archaea_march2021.csv")
functions<-functions[match(tree$tip.label,functions$X), ] #match the tree
rownames(functions)<-functions$X
functions$X<-NULL

sampleinfo<-read.csv("taxonomy/sample_names_ids_withotherfactors_tomedit.2.csv")
rownames(sampleinfo)<-sampleinfo$termite_tree_labels
sampleinfo<-sampleinfo[match(tree$tip.label,rownames(sampleinfo)), ] #match the tree

#run phylo.signal-<get signficiant p.values for phylogenetic correlation for each taxa>
library(adephylo)
library(ape)
library(phylobase)
library(phylosignal)

#clr transformed data with 0 values-<all funcs>
p4d <- phylo4d(tree,functions)
p2<-phyloSignal(p4d = p4d, method = "I")

#adjusted p-value using Benjamini & Hochberg method ("BH") or fdr method-
mod2 = as.data.frame(p2$pvalue)
mod2$padjusted_lambda<-p.adjust(mod2$I, method='fdr')
mod2$functions<-rownames(mod2)
mod2<-mod2%>%dplyr::mutate(significance_padjusted=ifelse(padjusted_lambda<0.05,"*",""))
mod2<-mod2%>%mutate(significance_I=ifelse(I<0.05,"*",""))

write.csv(mod2,file="phylogenetic_autocorrelation_mcr_methanogenesis_129samples_march2021.csv")
#--------------------------------------------------------------------------
#phyl.pca-
library(geomorph)
functions<-functions+100 #add a constant to every value


phyl.anova.func2_list<-function(filename){
filename<-as.data.frame(t(filename)) #convert row to column
filename$label<-rownames(filename)
geo_functions<-merge(filename,sampleinfo,by.x="label",by.y="termite_tree_labels")
geo_functions$diet_type_number4<-as.factor(geo_functions$diet_type_number4)
geo_functions[,2]<-as.numeric(geo_functions[,2])

gdf <- geomorph.data.frame(abundance = geo_functions[,2], diet = as.factor(geo_functions$diet_type_number4), phy = tree) #create a dataframe that contains the data and phylogeny. [according to google forum]

#group_name <- names(filename)[2] #get the correct column name. column 2 is the cazyme.

geomorph.lm<-procD.lm(abundance ~ diet, iter=99999,RRPP=TRUE,effect.type = "F", SS.type = "II" ,data=gdf,print.progress = FALSE) #SS.type="II" is used if the groups (i.e. diet/lineage) are of uneven sizes.
mod<-anova(geomorph.lm) #to examine if there is a phylogenetic correlation between "y" and "x"
mod2<-as.data.frame(mod$table)
mod2<-mod2[1,7]
# d is distance between vector lengths is the difference in the amount of shape change per unit of size. If d is 0.05 i.e., | 0.20 - 0.15 |.  Why do this?  Two vectors can be correlated in direction - the way shape changes with size - but one group can exhibit more shape change per unit size than the other
#D tells us the effect size of the standard deviation between two groups. Positive and above 0.2-0.3 means larger positive difference, while negative and below -0.2 to -0.3 means larger negative difference.

pw<-pairwise(geomorph.lm,groups = as.factor(gdf$diet))
pw_summary<-summary(pw,confidence=0.95,show.vectors = TRUE)
pw_summary2<-as.data.frame(pw_summary$summary.table)

pw_summary2$padjust<-p.adjust(pw_summary2$`Pr > d`,method="fdr") #fdr correction of p.value
pw_summary2$diets<-rownames(pw_summary2)

#convert the output to correct formats-
names<-colnames(geo_functions[2])
pw_summary2$genename<-rep(names,times=nrow(pw_summary2))
library(tidyr)
pw_summary3<-pw_summary2%>%dplyr::select("padjust","diets","genename")
pw_spread<-pw_summary3%>%spread(diets,padjust)
pw_spread2<-cbind(pw_spread,mod2)

return(pw_spread2)
}

functions_t<-as.data.frame(t(functions))
xy.list <-  split(functions_t, seq(nrow(functions_t))) #convert df to a list

kegg_geomorph<-lapply(xy.list,phyl.anova.func2_list) #apply on all columns

kegg_geomorph_df<-do.call(rbind.data.frame,kegg_geomorph)
write.csv(kegg_geomorph_df,file="Phylogenetic_anova_withdiet_mcr_methanogenesis_129samples_march2021.csv")
#---------------------------------------------------------------------------
##mcrABG in different diets-<clr-transformed>

methano_archaea<-read.csv("mcr_archaea_march2021.csv")

sampleinfo<-read.csv("taxonomy/sample_names_ids_withotherfactors_tomedit.2.csv")

methano_archaea_samples<-merge(methano_archaea,sampleinfo,by.x="X",by.y="termite_tree_labels")
setDT(methano_archaea_samples)
methano_diet<-methano_archaea_samples[, .(mcr = sum(mcr)), by = .(diet_type_number4)]
methano_diet$diet_type_number4<-gsub("^","diet_",methano_diet$diet_type_number4)
rownames(methano_diet)<-methano_diet$diet_type_number4

methano_diet
#----------------------------------------------------------------------------
##mcrABG in specific termite samples-
methano_archaea%>%arrange(desc(mcr))%>%head(10)




```


##mcrABG gene operons from metagenomes-
```{r}
##mcrABG-

functions<-read.csv("functions/all-metagenome-annotation-taxonomy-1000bps.txt")
functions$domain<-gsub("_p__.*$","",functions$taxa)

#get archaeal contigs out-
functions_archaea<-functions%>%filter(domain=="d__Archaea")

#get gene of interest out-

genes_vector<-as.vector(c("K00399","K00401","K00402"))

methano_data<-functions_archaea%>%filter(annotation %in% genes_vector) #condition1

#sum TPM per gene, per contig-
library(data.table)
setDT(methano_data)
genes<-methano_data[, .(TPM = sum(TPM)), by = .(sample_names,annotation,contig_names)]

contig_count<-genes%>%group_by(sample_names,contig_names)%>%count()
#contig_count2<-contig_count #for one kegg input gene
contig_count2<-contig_count%>%filter(n>1) 

#extract contigs with more than 1 methanogenesis genes-
methano_contigs<-methano_data%>%filter(sample_names %in% contig_count2$sample_names & contig_names %in% contig_count2$contig_names)
methano_contigs$sample_contigs<-paste(methano_contigs$sample_names,methano_contigs$contig_names,sep="__")
methano_contigs2<-methano_contigs%>%select(sample_contigs,annotation,TPM)

methano_contigs3<-unique(methano_contigs2) #if duplicates are present!
#spread the data-
library(tidyr)
#if unique combinations of keys are not present-
methano_contigs_spread<-methano_contigs3%>%group_by(sample_contigs,annotation)%>%mutate(grouped_id=row_number())%>%spread(annotation,TPM)%>%select(-grouped_id)

#add taxa name-
methano_contigs_spread$contig_names<-gsub("^.*__","",methano_contigs_spread$sample_contigs)
methano_contigs_spread$sample_names<-gsub("__.*$","",methano_contigs_spread$sample_contigs)

methano_data_sub<-methano_data%>%select(sample_names,contig_names,taxa)
methano_data_sub2<-unique(methano_data_sub)
methano_contigs_spread2<-merge(methano_contigs_spread,methano_data_sub2,by=c("contig_names","sample_names"))

#add termite information-
termites<-read.csv("taxonomy/sample_names_ids_withotherfactors_tomedit.2.csv")
methano_contigs_final<-merge(methano_contigs_spread2,termites,by.x="sample_names",by.y="samples")

#select non zero columns-
methano_contigs_final2<-methano_contigs_final %>% drop_na() #selecting all non-zero values
#methano_contigs_final2$phyla<-unlist(lapply(strsplit(as.character(methano_contigs_final2$taxa),split="_"),"[",2))
write.csv(methano_contigs_final2,file="methanogenesis_mcrABG_operon_contigs.csv")

#-----------------------------------------------------------------------------------------------
#stats-

#1.how many archaeal families and orders are present in the operon?
methano_contigs_final2$taxa_family<-gsub("_g__.*$","",methano_contigs_final2$taxa)
methano_contigs_final2$taxa_order<-gsub("_f__.*$","",methano_contigs_final2$taxa)

methano_contigs_final2%>%select(taxa_family)%>%group_by(taxa_family)%>%count()
methano_contigs_final2%>%select(taxa_order)%>%group_by(taxa_order)%>%count()

#2. how many termite species and lineages are present?
methano_contigs_final2%>%select(species_name)%>%group_by(species_name)%>%count()
methano_contigs_final2%>%select(lineage,higher_lower)%>%group_by(lineage,higher_lower)%>%count()

#3. methanogenesis by termite lineages
methano_contigs_final2%>%filter(higher_lower=="lower")


#4. how many contigs have marker-genes in them?
markergenes<-read.csv("taxonomy/all-cogs-taxonomy-contiginfo-tpm-filtered-corrected.csv",row.names=1)
markergenes<-markergenes%>%select(annotation,contig_names,samples)
markergenes$sample_contigs<-paste(markergenes$samples,markergenes$contig_names,sep="__")
setDT(markergenes)
markergenes2<-as.data.table(markergenes)[, toString(annotation), by = list(samples,contig_names,sample_contigs)] #get marker-genes per contig

methano_contigs_markers<-merge(methano_contigs_final2,markergenes2,by.x=c("sample_names","contig_names","sample_contigs"),by.y=c("samples","contig_names","sample_contigs"),all.x=TRUE)

```


##nifHDKENB and nifHDK gene operons from metagenomes-
```{r}
##nifHDKENB-

functions<-read.csv("functions/all-metagenome-annotation-taxonomy-5000bps.txt")
functions$domain<-gsub("_p__.*$","",functions$taxa)

#get gene of interest out-
genes_vector<-as.vector(c("K02586","K02591","K02588","K02587","K02592","K02585"))

methano_data<-functions%>%filter(annotation %in% genes_vector) #condition1

#sum TPM per gene, per contig-
library(data.table)
setDT(methano_data)
genes<-methano_data[, .(TPM = sum(TPM)), by = .(sample_names,annotation,contig_names)]

contig_count<-genes%>%group_by(sample_names,contig_names)%>%count()
#contig_count2<-contig_count #for one kegg input gene
contig_count2<-contig_count%>%filter(n>1) 

#extract contigs with more than 1 nitrogen fixation genes-
methano_contigs<-methano_data%>%filter(sample_names %in% contig_count2$sample_names & contig_names %in% contig_count2$contig_names)
methano_contigs$sample_contigs<-paste(methano_contigs$sample_names,methano_contigs$contig_names,sep="__")
methano_contigs2<-methano_contigs%>%select(sample_contigs,annotation,TPM)

methano_contigs3<-unique(methano_contigs2) #if duplicates are present!
#spread the data-
library(tidyr)
#if unique combinations of keys are not present-
methano_contigs_spread<-methano_contigs3%>%group_by(sample_contigs,annotation)%>%mutate(grouped_id=row_number())%>%spread(annotation,TPM)%>%select(-grouped_id)

#add taxa name-
methano_contigs_spread$contig_names<-gsub("^.*__","",methano_contigs_spread$sample_contigs)
methano_contigs_spread$sample_names<-gsub("__.*$","",methano_contigs_spread$sample_contigs)

methano_data_sub<-methano_data%>%select(sample_names,contig_names,taxa)
methano_data_sub2<-unique(methano_data_sub)
methano_contigs_spread2<-merge(methano_contigs_spread,methano_data_sub2,by=c("contig_names","sample_names"))

#add termite information-
termites<-read.csv("taxonomy/sample_names_ids_withotherfactors_tomedit.2.csv")
methano_contigs_final<-merge(methano_contigs_spread2,termites,by.x="sample_names",by.y="samples")

#select non zero columns-
methano_contigs_final2<-methano_contigs_final %>% drop_na() #selecting all non-zero values
methano_contigs_final2<-methano_contigs_final2%>%filter(taxa!="d__Bacteria")
write.csv(methano_contigs_final2,file="nifHDKENB_operon_metagenomecontigs.csv")

#-----------------------------------------------------------------------------------------------
#stats-

#1.how many bacterial families and orders are present in the operon?
methano_contigs_final2$taxa_phyla<-gsub("_c__.*$","",methano_contigs_final2$taxa)

methano_contigs_final2%>%select(taxa_phyla)%>%group_by(taxa_phyla)%>%count()
methano_contigs_final2%>%filter(taxa_phyla=="d__Bacteria_p__Spirochaetota")
methano_contigs_final2%>%filter(taxa_phyla=="d__Archaea_p__Methanobacteriota")

#2. how many termite species and lineages are present?
methano_contigs_final2%>%select(species_name)%>%group_by(species_name)%>%count()
methano_contigs_final2%>%select(lineage,higher_lower)%>%group_by(lineage,higher_lower)%>%count()

#3. methanogenesis by termite lineages
methano_contigs_final2%>%filter(higher_lower=="lower")


#4. how many contigs have marker-genes in them?
markergenes<-read.csv("taxonomy/all-cogs-taxonomy-contiginfo-tpm-filtered-corrected.csv",row.names=1)
markergenes<-markergenes%>%select(annotation,contig_names,samples)
markergenes$sample_contigs<-paste(markergenes$samples,markergenes$contig_names,sep="__")
setDT(markergenes)
markergenes2<-as.data.table(markergenes)[, toString(annotation), by = list(samples,contig_names,sample_contigs)] #get marker-genes per contig

methano_contigs_markers<-merge(methano_contigs_final2,markergenes2,by.x=c("sample_names","contig_names","sample_contigs"),by.y=c("samples","contig_names","sample_contigs"),all.x=TRUE)

##-------------------------------------------------------------------------------------------------
##-------------------------------------------------------------------------------------------------
#nifHDK-

##FROM 1000bps contigs-
functions<-read.csv("functions/all-metagenome-annotation-taxonomy-1000bps.txt")
functions$domain<-gsub("_p__.*$","",functions$taxa)

#get gene of interest out-
genes_vector<-as.vector(c("K02586","K02591","K02588"))

methano_data<-functions%>%filter(annotation %in% genes_vector) #condition1

#sum TPM per gene, per contig-
library(data.table)
setDT(methano_data)
genes<-methano_data[, .(TPM = sum(TPM)), by = .(sample_names,annotation,contig_names)]

contig_count<-genes%>%group_by(sample_names,contig_names)%>%count()
#contig_count2<-contig_count #for one kegg input gene
contig_count2<-contig_count%>%filter(n>1) 

#extract contigs with more than 1 nitrogen fixation genes-
methano_contigs<-methano_data%>%filter(sample_names %in% contig_count2$sample_names & contig_names %in% contig_count2$contig_names)
methano_contigs$sample_contigs<-paste(methano_contigs$sample_names,methano_contigs$contig_names,sep="__")
methano_contigs2<-methano_contigs%>%select(sample_contigs,annotation,TPM)

methano_contigs3<-unique(methano_contigs2) #if duplicates are present!
#spread the data-
library(tidyr)
#if unique combinations of keys are not present-
methano_contigs_spread<-methano_contigs3%>%group_by(sample_contigs,annotation)%>%mutate(grouped_id=row_number())%>%spread(annotation,TPM)%>%select(-grouped_id)

#add taxa name-
methano_contigs_spread$contig_names<-gsub("^.*__","",methano_contigs_spread$sample_contigs)
methano_contigs_spread$sample_names<-gsub("__.*$","",methano_contigs_spread$sample_contigs)

methano_data_sub<-methano_data%>%select(sample_names,contig_names,taxa)
methano_data_sub2<-unique(methano_data_sub)
methano_contigs_spread2<-merge(methano_contigs_spread,methano_data_sub2,by=c("contig_names","sample_names"))

#add termite information-
termites<-read.csv("taxonomy/sample_names_ids_withotherfactors_tomedit.2.csv")
methano_contigs_final<-merge(methano_contigs_spread2,termites,by.x="sample_names",by.y="samples")

#select non zero columns-
methano_contigs_final2<-methano_contigs_final %>% drop_na() #selecting all non-zero values
methano_contigs_final2<-methano_contigs_final2%>%filter(taxa!="d__Bacteria")
write.csv(methano_contigs_final2,file="nifHDK_operon_1000bpscontigs_metagenomecontigs.csv")


#4. how many contigs have marker-genes in them?
markergenes<-read.csv("taxonomy/all-cogs-taxonomy-contiginfo-tpm-filtered-corrected.csv",row.names=1)
markergenes<-markergenes%>%select(annotation,contig_names,samples)
markergenes$sample_contigs<-paste(markergenes$samples,markergenes$contig_names,sep="__")
setDT(markergenes)
markergenes2<-as.data.table(markergenes)[, toString(annotation), by = list(samples,contig_names,sample_contigs)] #get marker-genes per contig

methano_contigs_markers<-merge(methano_contigs_final2,markergenes2,by.x=c("sample_names","contig_names","sample_contigs"),by.y=c("samples","contig_names","sample_contigs"),all.x=TRUE)

#------------------------------------------------------------------------------------------------------
##FROM 5000BPS CONTIG-

functions<-read.csv("functions/all-metagenome-annotation-taxonomy-5000bps.txt")
functions$domain<-gsub("_p__.*$","",functions$taxa)

#get gene of interest out-
genes_vector<-as.vector(c("K02586","K02591","K02588"))

methano_data<-functions%>%filter(annotation %in% genes_vector) #condition1

#sum TPM per gene, per contig-
library(data.table)
setDT(methano_data)
genes<-methano_data[, .(TPM = sum(TPM)), by = .(sample_names,annotation,contig_names)]

contig_count<-genes%>%group_by(sample_names,contig_names)%>%count()
#contig_count2<-contig_count #for one kegg input gene
contig_count2<-contig_count%>%filter(n>1) 

#extract contigs with more than 1 nitrogen fixation genes-
methano_contigs<-methano_data%>%filter(sample_names %in% contig_count2$sample_names & contig_names %in% contig_count2$contig_names)
methano_contigs$sample_contigs<-paste(methano_contigs$sample_names,methano_contigs$contig_names,sep="__")
methano_contigs2<-methano_contigs%>%select(sample_contigs,annotation,TPM)

methano_contigs3<-unique(methano_contigs2) #if duplicates are present!
#spread the data-
library(tidyr)
#if unique combinations of keys are not present-
methano_contigs_spread<-methano_contigs3%>%group_by(sample_contigs,annotation)%>%mutate(grouped_id=row_number())%>%spread(annotation,TPM)%>%select(-grouped_id)

#add taxa name-
methano_contigs_spread$contig_names<-gsub("^.*__","",methano_contigs_spread$sample_contigs)
methano_contigs_spread$sample_names<-gsub("__.*$","",methano_contigs_spread$sample_contigs)

methano_data_sub<-methano_data%>%select(sample_names,contig_names,taxa)
methano_data_sub2<-unique(methano_data_sub)
methano_contigs_spread2<-merge(methano_contigs_spread,methano_data_sub2,by=c("contig_names","sample_names"))

#add termite information-
termites<-read.csv("taxonomy/sample_names_ids_withotherfactors_tomedit.2.csv")
methano_contigs_final<-merge(methano_contigs_spread2,termites,by.x="sample_names",by.y="samples")

#select non zero columns-
methano_contigs_final2<-methano_contigs_final %>% drop_na() #selecting all non-zero values
methano_contigs_final2<-methano_contigs_final2%>%filter(taxa!="d__Bacteria")
write.csv(methano_contigs_final2,file="nifHDK_operon_5000bpscontigs_metagenomecontigs.csv")

#4. how many contigs have marker-genes in them?
markergenes<-read.csv("taxonomy/all-cogs-taxonomy-contiginfo-tpm-filtered-corrected.csv",row.names=1)
markergenes<-markergenes%>%select(annotation,contig_names,samples)
markergenes$sample_contigs<-paste(markergenes$samples,markergenes$contig_names,sep="__")
setDT(markergenes)
markergenes2<-as.data.table(markergenes)[, toString(annotation), by = list(samples,contig_names,sample_contigs)] #get marker-genes per contig

methano_contigs_markers<-merge(methano_contigs_final2,markergenes2,by.x=c("sample_names","contig_names","sample_contigs"),by.y=c("samples","contig_names","sample_contigs"),all.x=TRUE)

#-------------------------------------------------------------------------------------
##NOTE- Number of nifHDK containing contigs in metagenomes above 1000bps and 5000bps is exactly the same. Number of marker-gene containing nifHDK contigs is also exactly the same (9 contigs out of 37 contigs). So, 5000bp contigs stats is used.

#stats-


methano_contigs_final2$taxa_phyla<-gsub("_c__.*$","",methano_contigs_final2$taxa)


#how many Bacteroidota in lower termites-
methano_contigs_final2%>%filter(higher_lower=="lower")%>%select(taxa_phyla)%>%group_by(taxa_phyla)%>%count()
bacteroidota_lower<-methano_contigs_final2%>%filter(higher_lower=="lower" & taxa_phyla=="d__Bacteria_p__Bacteroidota")

#how many Bacteroidota in higher termites-
methano_contigs_final2%>%filter(taxa_phyla=="d__Bacteria_p__Bacteroidota" & higher_lower=="higher")

#how many Bacteroidota in Heterotermes_Coptotermes-
methano_contigs_final2%>%filter(taxa_phyla=="d__Bacteria_p__Bacteroidota" & lineage2=="Heterotermes_Coptotermes")

#how many Bacteroidota in no Heterotermes_Coptotermes lower termites-
methano_contigs_final2%>%filter(higher_lower=="lower" & taxa_phyla=="d__Bacteria_p__Bacteroidota" & lineage2!="Heterotermes_Coptotermes")

#how many termite lineages in no Heterotermes_Coptotermes lower termites-
methano_contigs_final2%>%filter(higher_lower=="lower" & taxa_phyla=="d__Bacteria_p__Bacteroidota" & lineage2!="Heterotermes_Coptotermes")%>%select(lineage)%>%group_by(lineage)%>%count()

#which is the most dominant phyla in higher termites-
methano_contigs_final2%>%filter(higher_lower=="higher")%>%select(taxa_phyla)%>%group_by(taxa_phyla)%>%count()


#which is the most dominant phyla in Microcerotermes-
methano_contigs_final2%>%filter(lineage=="Microcerotermes")%>%select(taxa_phyla)%>%group_by(taxa_phyla)%>%count()

methano_contigs_final2%>%filter(lineage=="Microcerotermes")%>%select(taxa)

#which is the most dominant phyla in wood-feeding lower termites-
methano_contigs_final2%>%filter(diet_type_number4=="2")%>%select(taxa_phyla)%>%group_by(taxa_phyla)%>%count()
methano_contigs_final2%>%filter(diet_type_number4=="2" & taxa_phyla=="d__Bacteria_p__Fibrobacterota")

#taxa specific trends-
methano_contigs_final2%>%filter(taxa_phyla=="d__Bacteria_p__Bacteroidota" & higher_lower=="higher")

methano_contigs_final2%>%filter(taxa_phyla=="d__Bacteria_p__Spirochaetota" & higher_lower=="lower")
```


##table S11-
```{r}
functions<-read.csv("metabolic_pathways_ordered_clrtransformed_124samples_19genesnosubunits_feb2021.csv")
methano_archaea<-read.csv("methanogenesis_archaea_mcrsummed_feb2021.csv")

allmetabolicgenes<-merge(functions,methano_archaea,by="X")

sampleinfo<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv")

allmetabolicgenes_samples<-merge(allmetabolicgenes,sampleinfo,by.x="X",by.y="termite_tree_labels")
cols<-colnames(allmetabolicgenes)
cols<-cols[-1]

setDT(allmetabolicgenes_samples)
allmetabolicgenes_diet<-allmetabolicgenes_samples[, lapply(.SD, sum, na.rm=TRUE), by=diet_type_number4, .SDcols=cols ]
allmetabolicgenes_diet$diet_type_number4<-gsub("^","diet_",allmetabolicgenes_diet$diet_type_number4)
rownames(allmetabolicgenes_diet)<-allmetabolicgenes_diet$diet_type_number4


write.csv(allmetabolicgenes_diet,file="allmetabolicgenes-mcr-clr-perdiet-feb2021.csv")
```
