---
title: "metabolic_genes"
author: "Jigyasa_Arora"
date: "2/28/2021"
---

```{r}

library(dplyr)
library(tidyr)
library(ape)
library(data.table)
```

##Extracting out metabolic genes from contigs+summing by subunit+using Hydrogenases classified by Hyddb-
```{r}  
 
functions<-read.csv("functions_124samples_11158columns_rawtpm_feb2021.csv") ##generated in "functions.Rmd"


#remove the FeFe and NiFe hydrogenases catalytic subunit columns, as we are adding the FeFe and NiFe catalytic subunits from Hydd classification. otherwise it will be repeated. 
#remove the FeFe and NiFe hydrogenases catalytic subunit columns, as we are adding the FeFe and NiFe catalytic subunits from Hydd classification. otherwise it will be repeated. 
functions<-functions%>%select(-c(PF00374.19)) #"PF00374.19" is present as a single column
functions <- functions[, -grep(pattern = "PF02906.14", colnames(functions))] #"PF02906.14" is present with other PFAMids.

#add hydrogenases data-
nife<-read.csv("nife_hydd_groups_prokTPM.csv",row.names=1)
nife$V2<-gsub("\\[NiFe]\\ ","",nife$V2)
nife$classification<-gsub('.{1}$', '', nife$V2)
setDT(nife)
nife2<-nife[, .(TPM = sum(TPM)), by = .(samples,classification)]
nife_spread<-as.data.frame(spread(nife2,classification,TPM))

fefe<-read.csv("fefe_hydd_groups_prokTPM.csv",row.names=1)
fefe$V2<-gsub("\\[FeFe]\\ ","",fefe$V2)
fefe<-fefe%>%mutate(classification=ifelse(V2=="Group C2" | V2=="Group C3","Group C",fefe$V2))
setDT(fefe)
fefe2<-fefe[, .(TPM = sum(TPM)), by = .(samples,classification)]
fefe_spread<-as.data.frame(spread(fefe2,classification,TPM))

hydrogenases<-merge(nife_spread,fefe_spread,by="samples",all=TRUE)

sampleinfo<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv")
sampleinfo<-sampleinfo%>%select(samples,termite_tree_labels)

hydrogenases<-merge(hydrogenases,sampleinfo,by="samples")
write.csv(hydrogenases,file="functions-hydrogenases-hydddb-allsamples-feb2021.csv")

functions2<-merge(hydrogenases,functions,by.x="termite_tree_labels",by.y="X",all.y=TRUE)
functions2$samples<-NULL
functions2<-as.data.frame(functions2)
rownames(functions2)<-functions2$termite_tree_labels
#--------------------------------------------------------------------------------------
##summing by subunits-
main_funcs<-read.csv("allbacterial_koids_metabolicgenes_jan2021.csv",header=FALSE)

#1.
acscols<-c("K00198","K00194","K00197") #"K14138", "K15023" not found.
functions2$acs<-functions2%>%select("K00198","K00194","K00197")%>%rowSums(na.rm=TRUE)
functions2<-functions2%>%select(-c("K00198","K00194","K00197"))

#2.
apr<-c("K00394","K00395")
functions2$apr<-functions2%>%select("K00394","K00395")%>%rowSums(na.rm=TRUE)
functions2<-functions2%>%select(-c("K00394","K00395"))

#3.
nif<-c("K02586","K02591","K02588")
functions2$nif<-functions2%>%select("K02586","K02591","K02588")%>%rowSums(na.rm=TRUE)
functions2<-functions2%>%select(-c("K02586","K02591","K02588"))

#4. 
nar<-c("K00370","K00371","K00374")
functions2$nar<-functions2%>%select("K00370","K00371","K00374")%>%rowSums(na.rm=TRUE)
functions2<-functions2%>%select(-c("K00370","K00371","K00374"))

#5.
nap<-c("K02567","K02568")
functions2$nap<-functions2%>%select("K02567","K02568")%>%rowSums(na.rm=TRUE)
functions2<-functions2%>%select(-c("K02567","K02568"))

#6.
nir<-c("K00362","K00363")
functions2$nir<-functions2%>%select("K00362","K00363")%>%rowSums(na.rm=TRUE)
functions2<-functions2%>%select(-c("K00362","K00363"))

#7.
nrf<-c("K03385") #only nrfA present.K15876 absent
functions2$nrf<-functions2%>%select("K03385")%>%rowSums(na.rm=TRUE)
functions2<-functions2%>%select(-c("K03385"))

#8.
ure<-c("K01428","K01429","K01430")
functions2$ure<-functions2%>%select("K01428","K01429","K01430")%>%rowSums(na.rm=TRUE)
functions2<-functions2%>%select(-c("K01428","K01429","K01430"))

#9.
glt<-c("K00265","K00266")
functions2$glt<-functions2%>%select("K00265","K00266")%>%rowSums(na.rm=TRUE)
functions2<-functions2%>%select(-c("K00265","K00266"))

#10.
arg<-c("K00611","K01940","K01755")
functions2$arg<-functions2%>%select("K00611","K01940","K01755")%>%rowSums(na.rm=TRUE)
functions2<-functions2%>%select(-c("K00611","K01940","K01755"))

write.csv(functions2,file="allfunctions_rawTPM_124samples_genesnosubunit_hydrogenases.csv")

```


##Filters and clr transformation of data-
```{r} 
##Transformations-

##clr transform-
clr_transform<-function(filename_spread2){ 
library(propr) 
filename_spread2[is.na(filename_spread2)] <-0 #convert NA to zero
#rownames(filename_spread2)<-filename_spread2$label
filename_spread2$termite_tree_labels<-NULL
filename_spread2$samplename<-NULL
#add psuedo counts to account for missing and not-present data-
filename_spread2<-filename_spread2+0.65

filename_propr<-propr(filename_spread2,metric="rho",ivar="clr")

#creating a df with all the transformed pathways and metadata-
filename_propr_df<-filename_propr@logratio

return(filename_propr_df)
}

functions_clr<-clr_transform(functions2)
#rownames(functions_clr)<-functions2$label

##select annotations present in more than 10% of the samples-(13/124)
above10perc_function<-function(filename){
  filename$label<-NULL
  filename$samplename<-NULL
 
  filename2<-filename #for filtering
  filename2[filename2 > 0] <- 1
  filename2[filename2 <= 0] <-0
  filename_10perc<-filename2[colSums(filename2) > 13]
  columns<-colnames(filename_10perc)

  filename3<-filename%>%dplyr::select(columns)
  return(filename3)
}

functions_above10<-above10perc_function(functions_clr)

morethan3taxa_functions<-function(filename){

filename$label<-NULL
filename$samplename<-NULL
filename2<-filename
filename2[filename2 > 0] <- 1
filename2[filename2 <= 0] <-0
filename_threeabove<-filename2[rowSums(filename2[,])>=3, ]
rows<-rownames(filename_threeabove)

filename3<-subset(filename, rownames(filename) %in% rows)

return(filename3)
}

functions_above10_3above<-morethan3taxa_functions(functions_above10)
write.csv(functions_above10_3above,file="allfunctions_clr_124samples_genesnosubunit_hydrogenases.csv")


##select the kegg genes of interest-
main_funcs<-read.csv("allbacterial_koids_metabolicgenes_nosubunits_jan2021.csv",header=FALSE) #genes with subunits have correct names in this file

mainfuncs_functions<-functions_above10_3above%>%dplyr::select_if(names(.) %in% main_funcs$V1) #select the keggs out
mainfuncs_functions$termitenames<-rownames(mainfuncs_functions)

main_funcs$V1[!(main_funcs$V1) %in% colnames(mainfuncs_functions)] #find keggs not present

main_funcs2<-main_funcs%>%filter(V1 %in% c("Group 1","Group A","K00297","K00625","K00925","K00926","K01491","K01915","K01938","K03320","acs","apr","nif","nar","nap","nrf","ure","glt","arg")) #keep genes in metagenome contigs 
genenames<-as.vector(main_funcs2$V1)
mainfuncs_functions_ordered<-mainfuncs_functions[,genenames] #order the metagenome contigs data


tree<-read.tree("1000bps_taxonomy_124taxa_tree.nwk")
mainfuncs_functions_ordered<-mainfuncs_functions_ordered[match(tree$tip.label,rownames(mainfuncs_functions_ordered)), ]

write.csv(mainfuncs_functions_ordered,file="metabolic_pathways_ordered_clrtransformed_124samples_19genesnosubunits_feb2021.csv")
```

##Moran's I and Phyl.anova stats-
```{r}
#Moran's I-
functions<-read.csv("metabolic_pathways_ordered_clrtransformed_124samples_19genesnosubunits_feb2021.csv",row.names = 1) #clr transformed + summed by subunit
tree<-read.tree("1000bps_taxonomy_124taxa_tree.nwk")

functions<-functions[match(tree$tip.label,rownames(functions)), ] #match the tree



#run phylo.signal-<get signficiant p.values for phylogenetic correlation for each taxa>
library(adephylo)
library(ape)
library(phylobase)
library(phylosignal)

#clr transformed data with 0 values-<all funcs>
p4d <- phylo4d(tree,functions)
p2<-phyloSignal(p4d = p4d, method = "I")

#adjusted p-value using Benjamini & Hochberg method ("BH") or fdr method-
mod2 = as.data.frame(p2$pvalue)
mod2$padjusted_lambda<-p.adjust(mod2$I, method='fdr')
mod2$functions<-rownames(mod2)
mod2<-mod2%>%dplyr::mutate(significance_padjusted=ifelse(padjusted_lambda<0.05,"*",""))
mod2<-mod2%>%mutate(significance_I=ifelse(I<0.05,"*",""))
main_funcs<-read.csv("allbacterial_koids_metabolicgenes_nosubunits_jan2021.csv",header=FALSE)
main_funcs$V1<-gsub(" ",".",main_funcs$V1) #match the column with colnames of file.
mod2<-merge(mod2,main_funcs,by.x="functions",by.y="V1")

write.csv(mod2,file="phylogenetic_autocorrelation_19etabolicgenes_nosubunit_124samples_feb2021.csv")
#-------------------------------------------------------------------------------------------------------------
##phyl.anova-
library(geomorph)
functions<-functions+100 #add a constant to every value

sampleinfo<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv",row.names=1)
rownames(sampleinfo)<-sampleinfo$termite_tree_labels
sampleinfo<-sampleinfo[match(tree$tip.label,rownames(sampleinfo)), ] #match the tree

phyl.anova.func2_list<-function(filename){
filename<-as.data.frame(t(filename)) #convert row to column
filename$label<-rownames(filename)
geo_functions<-merge(filename,sampleinfo,by.x="label",by.y="termite_tree_labels")
geo_functions$diet_type_number4<-as.factor(geo_functions$diet_type_number4)
geo_functions[,2]<-as.numeric(geo_functions[,2])

gdf <- geomorph.data.frame(abundance = geo_functions[,2], diet = as.factor(geo_functions$diet_type_number4), phy = tree) #create a dataframe that contains the data and phylogeny. [according to google forum]

#group_name <- names(filename)[2] #get the correct column name. column 2 is the cazyme.

geomorph.lm<-procD.lm(abundance ~ diet, iter=99999,RRPP=TRUE,effect.type = "F", SS.type = "II" ,data=gdf,print.progress = FALSE) #SS.type="II" is used if the groups (i.e. diet/lineage) are of uneven sizes.
mod<-anova(geomorph.lm) #to examine if there is a phylogenetic correlation between "y" and "x"
mod2<-as.data.frame(mod$table)
mod2<-mod2[1,7]
# d is distance between vector lengths is the difference in the amount of shape change per unit of size. If d is 0.05 i.e., | 0.20 - 0.15 |.  Why do this?  Two vectors can be correlated in direction - the way shape changes with size - but one group can exhibit more shape change per unit size than the other
#D tells us the effect size of the standard deviation between two groups. Positive and above 0.2-0.3 means larger positive difference, while negative and below -0.2 to -0.3 means larger negative difference.

pw<-pairwise(geomorph.lm,groups = as.factor(gdf$diet))
pw_summary<-summary(pw,confidence=0.95,show.vectors = TRUE)
pw_summary2<-as.data.frame(pw_summary$summary.table)

pw_summary2$padjust<-p.adjust(pw_summary2$`Pr > d`,method="fdr") #fdr correction of p.value
pw_summary2$diets<-rownames(pw_summary2)

#convert the output to correct formats-
names<-colnames(geo_functions[2])
pw_summary2$genename<-rep(names,times=nrow(pw_summary2))
library(tidyr)
pw_summary3<-pw_summary2%>%dplyr::select("padjust","diets","genename")
pw_spread<-pw_summary3%>%spread(diets,padjust)
pw_spread2<-cbind(pw_spread,mod2)

return(pw_spread2)
}

functions_t<-as.data.frame(t(functions))
xy.list <-  split(functions_t, seq(nrow(functions_t))) #convert df to a list

kegg_geomorph<-lapply(xy.list,phyl.anova.func2_list) #apply on all columns

kegg_geomorph_df<-do.call(rbind.data.frame,kegg_geomorph)


#find p.adjust for anova-
kegg_geomorph_df$padjust<-p.adjust(kegg_geomorph_df$mod2)

kegg_geomorph_df<-kegg_geomorph_df%>%mutate(significant_anova=ifelse(padjust <=0.05,"*",""))

#join with genenames-
main_funcs<-read.csv("allbacterial_koids_metabolicgenes_nosubunits_jan2021.csv",header=FALSE)
main_funcs$V1<-gsub(" ",".",main_funcs$V1) #match the column with colnames of file.
kegg_geomorph_df2<-merge(kegg_geomorph_df,main_funcs,by.x="genename",by.y="V1")

write.csv(kegg_geomorph_df2,file="Phylogenetic_anova_withdiet_19metabolicpathwaygenes_nosubunit_124samples_feb2021.csv")



```

##Basic stats on metabolic genes-
```{r} 


functions2<-read.csv("allfunctions_rawTPM_124samples_genesnosubunit_hydrogenases.csv")
rownames(functions2)<-functions2$X

main_funcs<-read.csv("allbacterial_koids_metabolicgenes_nosubunits_jan2021.csv",header=FALSE) #genes with subunits have correct names in this file
main_funcs$V1<-gsub(" ",".",main_funcs$V1) #match the column with colnames of file.

mainfuncs_functions<-functions2%>%dplyr::select_if(names(.) %in% main_funcs$V1) #select the keggs out
mainfuncs_functions$X<-rownames(mainfuncs_functions)

##relative abundance of each gene across diets-<raw TPM>

sampleinfo<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv")
sampleinfo<-sampleinfo%>%select(termite_tree_labels,diet_type_number4)

mainfuncs_functions_diet<-merge(mainfuncs_functions,sampleinfo,by.x="X",by.y="termite_tree_labels")

#sum by diet-
cols<-colnames(mainfuncs_functions_diet)
cols<-cols[-1]
setDT(mainfuncs_functions_diet)
mainfuncs_functions_diet2<-mainfuncs_functions_diet[, lapply(.SD, sum, na.rm=TRUE), by=diet_type_number4, .SDcols=cols ] #for all the cazymes
mainfuncs_functions_diet2$diet_type_number4<-gsub("^","diet_",mainfuncs_functions_diet2$diet_type_number4)
mainfuncs_functions_diet2<-as.data.frame(mainfuncs_functions_diet2)
rownames(mainfuncs_functions_diet2)<-mainfuncs_functions_diet2$diet_type_number4


#checking the relative abundance of genes per diet-
mainfuncs_functions_diet2%>%select(K00297,K00925,K01491,K01938)
#metF_ec1.5.1.20, ack_ec2.7.2.1, folD_ec1.5.1.5_ec3.5.4.9, fhs_ec6.3.4.3

write.csv(mainfuncs_functions_diet2,file="functions-metabolicgenes-bydiet-feb2021.csv")
#---------------------------------------------------------------------------------
#what is the relative abundance of hydrogenases in the metagenomes?
hydrogenases<-read.csv("functions-hydrogenases-hydddb-allsamples-feb2021.csv")
hydrogenases[is.na(hydrogenases)] <-0

fefe_overallsum<-sum(hydrogenases$Group.A)+sum(hydrogenases$Group.B)+sum(hydrogenases$Group.C)
nife_overallsum<-sum(hydrogenases$Group.1)+sum(hydrogenases$Group.2)
hydrogenases_overallsum<-fefe_overallsum+nife_overallsum

print(paste0("Overall abundance of FeFe hydrogenases is ", fefe_overallsum/nife_overallsum , " fold more than NiFe hydrogenases"))


print(paste0("Overall abundance of GroupA FeFe hydrogenases is ", sum(hydrogenases$Group.A)/fefe_overallsum*100))
print(paste0("Overall abundance of GroupB FeFe hydrogenases is ", sum(hydrogenases$Group.B)/fefe_overallsum*100))
print(paste0("Overall abundance of GroupC FeFe hydrogenases is ", sum(hydrogenases$Group.C)/fefe_overallsum*100))

print(paste0("Overall abundance of Group1 NiFe hydrogenases is ", sum(hydrogenases$Group.1)/nife_overallsum*100))
print(paste0("Overall abundance of Group2 NiFe hydrogenases is ", sum(hydrogenases$Group.2)/nife_overallsum*100))


#across diets-
#checking the relative abundance of genes per diet-
mainfuncs_functions_diet2%>%select(Group.A,Group.1)

#NiFe sub-classification-
nife<-read.csv("nife_hydd_groups_prokTPM.csv",row.names=1)
nife$V2<-gsub("\\[NiFe]\\ ","",nife$V2)
nife$classification<-gsub('.{1}$', '', nife$V2)

nife_group1d<-nife%>%filter(V2=="Group 1d")
sum(nife_group1d$TPM)/sum(nife$TPM)*100

print(paste0("Percent distribution of NiFe Group 1d in metagenomes are ", sum(nife_group1d$TPM)/sum(nife$TPM)*100, "%"))

#----------------------------------------------------------------------
##how many samples are sulfate reducing genes present?

functions2<-read.csv("allfunctions_rawTPM_124samples_genesnosubunit_hydrogenases.csv")
rownames(functions2)<-functions2$X

main_funcs<-read.csv("allbacterial_koids_metabolicgenes_nosubunits_jan2021.csv",header=FALSE) #genes with subunits have correct names in this file
main_funcs$V1<-gsub(" ",".",main_funcs$V1) #match the column with colnames of file.

mainfuncs_functions<-functions2%>%dplyr::select_if(names(.) %in% main_funcs$V1) #select the keggs out
mainfuncs_functions$X<-rownames(mainfuncs_functions)

mainfuncs_functions_sulfate<-mainfuncs_functions%>%select(apr,K00958)

#convert everything to binary-
mainfuncs_functions_sulfate_binary<-mainfuncs_functions_sulfate
mainfuncs_functions_sulfate_binary[is.na(mainfuncs_functions_sulfate_binary)] <-0 #convert NA to zero
mainfuncs_functions_sulfate_binary[mainfuncs_functions_sulfate_binary > 0] <- 1 
mainfuncs_functions_sulfate_binary[mainfuncs_functions_sulfate_binary <= 0] <-0

mainfuncs_functions_sulfate_binary_rows<-mainfuncs_functions_sulfate_binary %>% mutate(sumVar = rowSums(.[1:2])) #get the sum of two genes
rownames(mainfuncs_functions_sulfate_binary_rows)<-rownames(mainfuncs_functions_sulfate)
mainfuncs_functions_sulfate_binary_rows<-mainfuncs_functions_sulfate_binary_rows%>%select(sumVar)
mainfuncs_functions_sulfate_binary_rows[mainfuncs_functions_sulfate_binary_rows > 0] <- 1  #presence/absence of apr and sat genes in all termite samples.

print(paste0("sulfate reducing genes aprAB and sat are found in ",sum(mainfuncs_functions_sulfate_binary_rows$sumVar)/nrow(mainfuncs_functions_sulfate_binary_rows)*100, "% of termite samples"))

#----------------------------------------------------------------------------------------
##how many samples are nifHDK genes present?

functions2<-read.csv("allfunctions_rawTPM_124samples_genesnosubunit_hydrogenases.csv")
rownames(functions2)<-functions2$X

mainfuncs_functions_nif<-functions2%>%select(nif)
mainfuncs_functions_nif$X<-rownames(mainfuncs_functions_nif)

#what is the difference of nif genes across diet?
sampleinfo<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv")
sampleinfo<-sampleinfo%>%select(termite_tree_labels,diet_type_number4)

mainfuncs_functions_nif_diet<-merge(mainfuncs_functions_nif,sampleinfo,by.x="X",by.y="termite_tree_labels")
setDT(mainfuncs_functions_nif_diet)
mainfuncs_functions_nif_diet2<-mainfuncs_functions_nif_diet[, .(nif = sum(nif)), by = .(diet_type_number4)]

lt_wf<-mainfuncs_functions_nif_diet2[1,2]+mainfuncs_functions_nif_diet2[2,2]
lt<-mainfuncs_functions_nif_diet2[1,2]
wf<-mainfuncs_functions_nif_diet2[2,2]
sf<-mainfuncs_functions_nif_diet2[4,2]
fc<-mainfuncs_functions_nif_diet2[3,2]

print(paste0("LT+WF are ", lt_wf/sf ," fold abundant than SF"))
print(paste0("LT+WF are ", lt_wf/fc ," fold abundant than SF"))
print(paste0("LT is ", lt/sf ," fold abundant than SF"))
print(paste0("LT is ", lt/fc ," fold abundant than FC"))
print(paste0("WF is ", wf/sf ," fold abundant than SF"))
print(paste0("WF is ", wf/fc ," fold abundant than FC"))

```




##methanogenesis in metagenomes genes- 
```{r} 
library(dplyr)
library(ape)
library(data.table)
library(tidyr)
library(stringr)

tree<-read.tree("1000bps_taxonomy_124taxa_tree.nwk")
d<-data.frame(label=tree$tip.label)
d$samples<-unlist(lapply(strsplit(as.character(d$label),split="-"),"[",1))
d$samples<-gsub("_","-",d$samples)

#get the taxonomy of each annotation out-
taxonomy<-read.csv("uniq-all-metagenome-annotation-taxonomy-1000bps_feb2021.txt")
taxonomy$phyla<-unlist(lapply(strsplit(as.character(taxonomy$V1.y),split="\\|"),"[",1))
taxonomy$phyla<-gsub("p__","",taxonomy$phyla)

#get domain level classification-
alldomains<-read.csv("/bucket/BourguignonU/Jigs_backup/working_files/AIMS/paper1/markergenes/markers-rpkm/individualanalysis_feb2021/gtdb_bac120_ar122_domaininfo.csv")

taxonomy2<-merge(taxonomy,alldomains,by="phyla")


#sum by annotation tpm per sample-
setDT(taxonomy2)
taxonomy_functions<-taxonomy2[, .(TPM = sum(prokTPM)), by = .(samples,annotation,domain)]
#spread by annotation & domain-
taxonomy_functions$domain_annotation<-paste(taxonomy_functions$domain,taxonomy_functions$annotation,sep="_")
taxonomy_functions2<-taxonomy_functions%>%select(samples,TPM,domain_annotation)
taxonomy_functions_spread<-as.data.frame(spread(taxonomy_functions2,domain_annotation,TPM))

taxonomy_functions_spread_merged<-merge(taxonomy_functions_spread,d,by.x="samples",by.y="samples",all.y=TRUE)

#reorder the table to match the termite tree-
taxonomy_functions_spread_merged<-taxonomy_functions_spread_merged[match(tree$tip.label,taxonomy_functions_spread_merged$label), ]
rownames(taxonomy_functions_spread_merged)<-taxonomy_functions_spread_merged$label

write.csv(taxonomy_functions_spread_merged,file="functions-domainlevel-124samples-rawTPM-feb2021.csv")


```

##filters and clr-transformation of methanogenesis data-
```{r}
functions<-read.csv("functions-domainlevel-124samples-rawTPM-feb2021.csv",row.names=1) 

#clr transform the data-
clr_transform<-function(filename_spread2){ 
library(propr) 
filename_spread2[is.na(filename_spread2)] <-0 #convert NA to zero
#rownames(filename_spread2)<-filename_spread2$label
filename_spread2$label<-NULL
filename_spread2$samples<-NULL
#add psuedo counts to account for missing and not-present data-
filename_spread2<-filename_spread2+0.65

filename_propr<-propr(filename_spread2,metric="rho",ivar="clr")

#creating a df with all the transformed pathways and metadata-
filename_propr_df<-filename_propr@logratio

return(filename_propr_df)
}

functions_clr<-clr_transform(functions)
#rownames(functions_clr)<-functions$label

#get archaeal annotation out-
functions_clr_archaea<-functions_clr[grepl("Archaea", names(functions_clr))]


write.csv(functions_clr_archaea,file="functions-archaeal-124samples-clr-feb2021.csv")
#---------------------------------------------------------------------------------------------
#methanogenesis genes to extract from clr transformed data-
methano_genenames<-read.csv("allarchaeal_koids_metabolicgenes_jan2021.csv",header=FALSE)
methano_genenames$V1<-gsub("^","Archaea_",methano_genenames$V1)

methanofunctions<-functions_clr_archaea%>%dplyr::select_if(names(.) %in% methano_genenames$V1)

#--------------------------------------------------------------------------------------------

##NOTE- Not performing any filtering as mcrABG genes separetly are not present in all samples after filtering-
above10perc_function<-function(filename){
filename$label<-NULL   
filename$samplename<-NULL   
 
filename2<-filename #for filtering 
filename2[filename2 > 0] <- 1 
filename2[filename2 <= 0] <-0
filename_10perc<-filename2[colSums(filename2) > 13]  
columns<-colnames(filename_10perc) 

filename3<-filename%>%dplyr::select(columns)
return(filename3)
}

#from clr transformed data-> re-select rows with more than 3 functions 
morethan3taxa_functions<-function(filename){
 
filename$label<-NULL 
filename$samplename<-NULL
filename2<-filename
filename2[filename2 > 0] <- 1
filename2[filename2 <= 0] <-0
filename_threeabove<-filename2[rowSums(filename2[,])>=3, ]
rows<-rownames(filename_threeabove)

filename3<-subset(filename, rownames(filename) %in% rows)

return(filename3)
}


#10% of methanogenesis genes-
methanofunctions_clr_10perc<-above10perc_function(methanofunctions)
#more than 3 taxa per functions-
methanofunctions_clr_10perc_3above<-morethan3taxa_functions(methanofunctions_clr_10perc)

#------------------------------------------------------------------------------------------

#1-sum all mcrABG genes together-
mcr<-c("K00399","K00401","K00402")
methanofunctions_mcr<-methanofunctions%>%select("Archaea_K00399","Archaea_K00401","Archaea_K00402")%>%rowSums(na.rm=TRUE)%>%as.data.frame()
colnames(methanofunctions_mcr)<-c("mcr")
write.csv(methanofunctions_mcr,file="methanogenesis_archaea_mcrsummed_feb2021.csv")


```

##stats on methanogenesis data-
```{r}
##tests-

tree<-read.tree("1000bps_taxonomy_124taxa_tree.nwk")

functions<-read.csv("methanogenesis_archaea_mcrsummed_feb2021.csv")
functions<-functions[match(tree$tip.label,functions$X), ] #match the tree
rownames(functions)<-functions$X
functions$X<-NULL

sampleinfo<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv")
rownames(sampleinfo)<-sampleinfo$termite_tree_labels
sampleinfo<-sampleinfo[match(tree$tip.label,rownames(sampleinfo)), ] #match the tree

#run phylo.signal-<get signficiant p.values for phylogenetic correlation for each taxa>
library(adephylo)
library(ape)
library(phylobase)
library(phylosignal)

#clr transformed data with 0 values-<all funcs>
p4d <- phylo4d(tree,functions)
p2<-phyloSignal(p4d = p4d, method = "I")

#adjusted p-value using Benjamini & Hochberg method ("BH") or fdr method-
mod2 = as.data.frame(p2$pvalue)
mod2$padjusted_lambda<-p.adjust(mod2$I, method='fdr')
mod2$functions<-rownames(mod2)
mod2<-mod2%>%dplyr::mutate(significance_padjusted=ifelse(padjusted_lambda<0.05,"*",""))
mod2<-mod2%>%mutate(significance_I=ifelse(I<0.05,"*",""))

#--------------------------------------------------------------------------
#phyl.pca-
library(geomorph)
functions<-functions+100 #add a constant to every value


phyl.anova.func2_list<-function(filename){
filename<-as.data.frame(t(filename)) #convert row to column
filename$label<-rownames(filename)
geo_functions<-merge(filename,sampleinfo,by.x="label",by.y="termite_tree_labels")
geo_functions$diet_type_number4<-as.factor(geo_functions$diet_type_number4)
geo_functions[,2]<-as.numeric(geo_functions[,2])

gdf <- geomorph.data.frame(abundance = geo_functions[,2], diet = as.factor(geo_functions$diet_type_number4), phy = tree) #create a dataframe that contains the data and phylogeny. [according to google forum]

#group_name <- names(filename)[2] #get the correct column name. column 2 is the cazyme.

geomorph.lm<-procD.lm(abundance ~ diet, iter=99999,RRPP=TRUE,effect.type = "F", SS.type = "II" ,data=gdf,print.progress = FALSE) #SS.type="II" is used if the groups (i.e. diet/lineage) are of uneven sizes.
mod<-anova(geomorph.lm) #to examine if there is a phylogenetic correlation between "y" and "x"
mod2<-as.data.frame(mod$table)
mod2<-mod2[1,7]
# d is distance between vector lengths is the difference in the amount of shape change per unit of size. If d is 0.05 i.e., | 0.20 - 0.15 |.  Why do this?  Two vectors can be correlated in direction - the way shape changes with size - but one group can exhibit more shape change per unit size than the other
#D tells us the effect size of the standard deviation between two groups. Positive and above 0.2-0.3 means larger positive difference, while negative and below -0.2 to -0.3 means larger negative difference.

pw<-pairwise(geomorph.lm,groups = as.factor(gdf$diet))
pw_summary<-summary(pw,confidence=0.95,show.vectors = TRUE)
pw_summary2<-as.data.frame(pw_summary$summary.table)

pw_summary2$padjust<-p.adjust(pw_summary2$`Pr > d`,method="fdr") #fdr correction of p.value
pw_summary2$diets<-rownames(pw_summary2)

#convert the output to correct formats-
names<-colnames(geo_functions[2])
pw_summary2$genename<-rep(names,times=nrow(pw_summary2))
library(tidyr)
pw_summary3<-pw_summary2%>%dplyr::select("padjust","diets","genename")
pw_spread<-pw_summary3%>%spread(diets,padjust)
pw_spread2<-cbind(pw_spread,mod2)

return(pw_spread2)
}

functions_t<-as.data.frame(t(functions))
xy.list <-  split(functions_t, seq(nrow(functions_t))) #convert df to a list

kegg_geomorph<-lapply(xy.list,phyl.anova.func2_list) #apply on all columns

kegg_geomorph_df<-do.call(rbind.data.frame,kegg_geomorph)

#---------------------------------------------------------------------------
##mcrABG in different diets-<clr-transformed>

methano_archaea<-read.csv("methanogenesis_archaea_mcrsummed_feb2021.csv")

sampleinfo<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv")

methano_archaea_samples<-merge(methano_archaea,sampleinfo,by.x="X",by.y="termite_tree_labels")
setDT(methano_archaea_samples)
methano_diet<-methano_archaea_samples[, .(mcr = sum(mcr)), by = .(diet_type_number4)]
methano_diet$diet_type_number4<-gsub("^","diet_",methano_diet$diet_type_number4)
rownames(methano_diet)<-methano_diet$diet_type_number4

methano_diet
#----------------------------------------------------------------------------
##mcrABG in specific termite samples-
methano_archaea%>%arrange(desc(mcr))%>%head(10)




```

##table S11-
```{r}
functions<-read.csv("metabolic_pathways_ordered_clrtransformed_124samples_19genesnosubunits_feb2021.csv")
methano_archaea<-read.csv("methanogenesis_archaea_mcrsummed_feb2021.csv")

allmetabolicgenes<-merge(functions,methano_archaea,by="X")

sampleinfo<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv")

allmetabolicgenes_samples<-merge(allmetabolicgenes,sampleinfo,by.x="X",by.y="termite_tree_labels")
cols<-colnames(allmetabolicgenes)
cols<-cols[-1]

setDT(allmetabolicgenes_samples)
allmetabolicgenes_diet<-allmetabolicgenes_samples[, lapply(.SD, sum, na.rm=TRUE), by=diet_type_number4, .SDcols=cols ]
allmetabolicgenes_diet$diet_type_number4<-gsub("^","diet_",allmetabolicgenes_diet$diet_type_number4)
rownames(allmetabolicgenes_diet)<-allmetabolicgenes_diet$diet_type_number4


write.csv(allmetabolicgenes_diet,file="allmetabolicgenes-mcr-clr-perdiet-feb2021.csv")
```

