This folder contains all the Rscripts to reproduce the statistical analysis of the paper. The files include-
1. basicstats.Rmd - Basic stats of assembly and read mapping. All values mentioned in the paper. Also https://github.com/Jigyasa3/termite_146guts_microbes_function/blob/main/statistical_analysis/basicstats_march2021.html
2. taxonomy.Rmd - Marker-gene phylogenetic signal and phyl.anova signal with termite host phylogeny and diet respectively. Comparison of marker-gene taxonomy with 16S data.
3. functions.Rmd - CAZYmes' phylogenetic signal and phyl.anova signal with termite host phylogeny and diet respectively.
4. metabolicgenes.Rmd - Metabolic genes' phylogenetic signal and phyl.anova signal with termite host phylogeny and diet respectively. (eg- reductive acetogenesis, hydrogenases catalytic subunits, methanogenesis mcrABG genes, sulfate reduction genes, nitrogen recycling and nitrogen fixation genes). The KEGGids are in the Rmd file.
5. functional_taxonomy_cazymes.Rmd - Microbes performing CAZYmes' in contigs >5000bps, phylogenetic signal and phyl.anova signal with termite host phylogeny and diet respectively.
6. paper_figures.Rmd - Scripts to reproduce the figures. The output from this Rmd file were loaded to ITOL (https://itol.embl.de/login.cgi) and manually improved in Inkscape.

https://maggimars.github.io/AcanthareaPhotosymbiosis/Analysis.html
