---
title: "taxonomy analysis"
author: "Jigyasa_Arora"
date: "2/27/2021"
---


```{r}
library(dplyr)
library(ape)

```

## taxonomic analysis by marker gene data- Do all marker genes present on the same contig have similar taxonomic annotation?
```{r}
  
taxonomy_ver2<-read.csv("tpm_cogs_allsamples_feb2021_1000bpscontigs_100counts1pm_similartocontigs.csv") #markers present in contigs >1000 bps,>100 read count and >1TPM, similar to gtdb contig taxonomy.
taxonomy_ver2$X.1<-NULL
taxonomy_ver2$X.2<-NULL
taxonomy_ver2$X.x.1<-NULL
taxonomy_ver2$X<-NULL
taxonomy_ver2$X.x<-NULL
taxonomy_ver2$X.y<-NULL
taxonomy_ver2$X.y.1<-NULL
taxonomy_ver2$V1.y<-NULL
taxonomy_ver2$phyla.y<-NULL
taxonomy_ver2$class.y<-NULL
taxonomy_ver2$order.y<-NULL

##do we duplicates in COG protein annotations? 
taxonomy_ver2_proteins<-taxonomy_ver2%>%select(fullproteinnames)%>%unique()
nrow(taxonomy_ver2)
nrow(taxonomy_ver2_proteins)

print(paste0("There are no duplicates in the data as no. of proteins is same in both files ",nrow(taxonomy_ver2) , " = ", nrow(taxonomy_ver2_proteins)))

##do markers present on the same contig have similar taxonomic annotation?
taxonomy_contigs<-taxonomy_ver2%>%select(fullnames)%>%group_by(fullnames)%>%count()

print(paste0(taxonomy_contigs%>%filter(n>1)%>%nrow()/nrow(taxonomy_contigs)*100, "% of markers are present on the same contig"))


taxonomy_contigs2<-taxonomy_contigs%>%filter(n==2)
print(paste0(nrow(taxonomy_contigs2)/taxonomy_contigs%>%filter(n>1)%>%nrow()*100, "% of 11.38% have two markers per contigs"))
print(paste0(taxonomy_contigs%>%filter(n==3)%>%nrow()/taxonomy_contigs%>%filter(n>1)%>%nrow()*100, "% of 11.38% have three markers per contigs"))
print(paste0(taxonomy_contigs%>%filter(n==4)%>%nrow()/taxonomy_contigs%>%filter(n>1)%>%nrow()*100, "% of 11.38% have four markers per contigs"))
print(paste0(taxonomy_contigs%>%filter(n==5)%>%nrow()/taxonomy_contigs%>%filter(n>1)%>%nrow()*100, "% of 11.38% have five markers per contigs"))

##Example-for contigs with 2 marker genes-
taxonomy_ver2_contigtaxa<-taxonomy_ver2%>%select(fullnames,markertaxonomy)%>%filter(fullnames %in% taxonomy_contigs2$fullnames) #extract markers present on same contigs

taxonomy_ver2_contigtaxa$markertaxonomy<-gsub("\\|","_",taxonomy_ver2_contigtaxa$markertaxonomy)
taxonomy_ver2_contigtaxa$phyla<-gsub("_c__.*$","",taxonomy_ver2_contigtaxa$markertaxonomy)

taxonomy_ver2_contigtaxa$class<-gsub("_o__.*$","",taxonomy_ver2_contigtaxa$markertaxonomy)

taxonomy_ver2_contigtaxa$order<-gsub("_f__.*$","",taxonomy_ver2_contigtaxa$markertaxonomy)

taxonomy_ver2_contigtaxa$family<-gsub("_g__.*$","",taxonomy_ver2_contigtaxa$markertaxonomy)

#taxonomy_ver2_contigtaxa$genus<-gsub("_s__.*$","",taxonomy_ver2_contigtaxa$markertaxonomy)

library(data.table)
setDT(taxonomy_ver2_contigtaxa)
taxonomy_contigs_family<-as.data.table(taxonomy_ver2_contigtaxa)[, toString(family), by = list(fullnames)]
taxonomy_contigs_family2<-setDT(taxonomy_contigs_family)[, paste0("family", 1:2) := tstrsplit(V1, ",")]


head(taxonomy_contigs_family2,5) 

taxonomy_contigs_family2$family1<-gsub(" ","",taxonomy_contigs_family2$family1)
taxonomy_contigs_family2$family2<-gsub(" ","",taxonomy_contigs_family2$family2)

taxonomy_contigs_family2<-taxonomy_contigs_family2%>%mutate(sametaxonomy=ifelse(family1==family1,"TRUE","FALSE"))
taxonomy_contigs_family2%>%filter(sametaxonomy=="TRUE")%>%nrow() 

print(paste0( taxonomy_contigs_family2%>%filter(sametaxonomy=="TRUE")%>%nrow()/nrow(taxonomy_contigs_family2)*100, "% of two markers have similar taxonomic annotation at family level when they are present on the same contig."))

```


## get marker gene taxonomy present in 124 samples-
```{r}

#get family level taxonomy of each marker gene-
taxonomy_ver2<-read.csv("tpm_cogs_allsamples_feb2021_1000bpscontigs_100counts1pm_similartocontigs.csv") #markers present in contigs >1000 bps,>100 read count and >1TPM, similar to gtdb contig taxonomy.
taxonomy_ver2$X.1<-NULL
taxonomy_ver2$X.2<-NULL
taxonomy_ver2$X.x.1<-NULL
taxonomy_ver2$X<-NULL
taxonomy_ver2$X.x<-NULL
taxonomy_ver2$X.y<-NULL
taxonomy_ver2$X.y.1<-NULL
taxonomy_ver2$V1.y<-NULL
taxonomy_ver2$phyla.y<-NULL
taxonomy_ver2$class.y<-NULL
taxonomy_ver2$order.y<-NULL

#taxonomy_ver2$markertaxonomy<-gsub("\\|","_",taxonomy_ver2$markertaxonomy)
#taxonomy_ver2$family<-gsub("_g__.*$","",taxonomy_ver2$markertaxonomy)

#spread the data by samples and taxonomy-
library(tidyr)
library(data.table)

setDT(taxonomy_ver2)
taxonomy_ver2_markertaxa<-taxonomy_ver2[, .(prokTPM = sum(prokTPM)), by = .(file_name,family)]

taxonomy_ver2_spread<-as.data.frame(spread(taxonomy_ver2_markertaxa,family,prokTPM))
rownames(taxonomy_ver2_spread)<-taxonomy_ver2_spread$file_name

#join the spread file with 124 termite tree to get 124 samples out-
#these 124 samples contain 10,000 contigs above 1000 bps, with more than 100 prokaryotic contigs.

newtree_124samples<-read.tree("1000bps_taxonomy_124taxa_tree.nwk")
dnew2<-data.frame(label=newtree_124samples$tip.label)

dnew2$label<-as.character(dnew2$label)
dnew2$runnumber<-unlist(lapply(strsplit(dnew2$label,split="-"),"[",1))
dnew2$runnumber<-gsub("_","-",dnew2$runnumber)


taxonomy_ver2_spread2<-merge(taxonomy_ver2_spread,dnew2,by.x="file_name",by.y="runnumber",all.y=TRUE)
rownames(taxonomy_ver2_spread2)<-taxonomy_ver2_spread2$label

##<to check from here>
#order the rownames by termite tree to get samples with no data at this step!
taxonomy_ver2_spread2<-taxonomy_ver2_spread2[match(newtree_124samples$tip.label,rownames(taxonomy_ver2_spread2)), ]
taxonomy_ver2_spread2[is.na(taxonomy_ver2_spread2)] <-0
write.csv(taxonomy_ver2_spread2,file="metagenome-taxonomy-1000bps-bymarkersspread-rawTPM-feb2021.csv")


head(taxonomy_ver2_spread2,5)
```



## data filtering for statistical analysis + clr transformation
```{r}

#clr transform the spread data-
taxonomy_ver2_spread2<-read.csv("metagenome-taxonomy-1000bps-bymarkersspread-rawTPM-feb2021.csv",row.names = 1)


clr_transform<-function(filename_spread2){ 
library(propr)
filename_spread2[is.na(filename_spread2)] <-0 #convert NA to zero
#rownames(filename_spread2)<-filename_spread2$label
filename_spread2$label<-NULL
filename_spread2$file_name<-NULL
#add psuedo counts to account for missing and not-present data-
filename_spread2<-filename_spread2+0.65

filename_propr<-propr(filename_spread2,metric="rho",ivar="clr")

#creating a df with all the transformed pathways and metadata-
filename_propr_df<-filename_propr@logratio

return(filename_propr_df)
}

markergene_taxonomy_124samples_clr<-clr_transform(taxonomy_ver2_spread2)




write.csv(markergene_taxonomy_124samples_clr,file="metagenome-taxonomy-1000bps-bymarkersspread-clr-feb2021.csv")

##check-
head(round(sum(markergene_taxonomy_124samples_clr[1,])),3) #the overall sum per sample (i.e row) should be close to zero after clr transformation.

nrow(markergene_taxonomy_124samples_clr)
ncol(markergene_taxonomy_124samples_clr)

##filtering the taxonomy file to find atleast 3 microbes in >5% of termite samples. 3 taxa cutoff is based on https://bioconductor.org/packages/release/bioc/vignettes/philr/inst/doc/philr-intro.html. While 10% of samples is an arbitrary cut-off as different metagenome analysis pipeline has different cut-offs. I wanted to make sure that we can capture data present in most of the termite lineage (for example, Apicotermitinae, soil-feeding Termitidae are not sequenced too deeply). 



#from clr transformed data->re-select columns present in 5% of samples (7/124)
above5perc_function<-function(filename){
filename$label<-NULL  
filename$samplename<-NULL  

filename2<-filename #for filtering 
filename2[filename2 > 0] <- 1 
filename2[filename2 <= 0] <-0
filename_10perc<-filename2[colSums(filename2) > 7]  
columns<-colnames(filename_10perc) 

filename3<-filename%>%dplyr::select(columns)
return(filename3)
}

markergene_taxonomy_124samples_clr_10perc<-above10perc_function(markergene_taxonomy_124samples_clr)
markergene_taxonomy_124samples_clr_5perc<-above5perc_function(markergene_taxonomy_124samples_clr)

#from clr transformed data-> re-select rows with more than 3 functions 
morethan3taxa_functions<-function(filename){
 
filename$label<-NULL 
filename$samplename<-NULL
filename2<-filename
filename2[filename2 > 0] <- 1
filename2[filename2 <= 0] <-0
filename_threeabove<-filename2[rowSums(filename2[,])>=3, ]
rows<-rownames(filename_threeabove)

filename3<-subset(filename, rownames(filename) %in% rows)

return(filename3)
}

markergene_taxonomy_124samples_clr_10perc_3above<-morethan3taxa_functions(markergene_taxonomy_124samples_clr_10perc)
markergene_taxonomy_124samples_clr_5perc_3above<-morethan3taxa_functions(markergene_taxonomy_124samples_clr_5perc)

write.csv(markergene_taxonomy_124samples_clr_5perc_3above,file="taxonomy_107samples_67taxa_feb2021.csv")
#-------------------------------------------------
##new tree-
newtree_124samples<-read.tree("1000bps_taxonomy_124taxa_tree.nwk")
newtree_107samples<-keep.tip(newtree_124samples,as.vector(rownames(markergene_taxonomy_124samples_clr_5perc_3above)))

write.tree(newtree_107samples,file="1000bps_taxonomy_107taxonomy_taxa_tree.nwk")
```

## taxonomy stats-Moran's I and Phyl.anova-
```{r}

tree<-read.tree("1000bps_taxonomy_107taxonomy_taxa_tree.nwk")

functions<-read.csv("taxonomy_107samples_67taxa_feb2021.csv",row.names = 1,header=TRUE)
functions$X<-NULL
functions<-functions[match(tree$tip.label,rownames(functions)), ] #match with the tree

library(adephylo)
library(ape) 
library(phylobase) 
library(phylosignal) 
 

functions[is.na(functions)] <-0 #convert any NA to zero
functions<-functions[match(tree$tip.label,rownames(functions)), ]
p4d <- phylo4d(tree,functions)
p2<-phyloSignal(p4d = p4d, method = "I")

#adjusted p-value using  fdr method-
mod2 = as.data.frame(p2$pvalue)
mod2$padjusted_lambda<-p.adjust(mod2$I, method='fdr')
mod2$functions<-rownames(mod2)
mod2<-mod2%>%dplyr::mutate(significance_padjusted=ifelse(padjusted_lambda<0.05,"*",""))
mod2<-mod2%>%dplyr::mutate(significance_moranI=ifelse(I<0.05,"*",""))

write.csv(mod2,file="phylogenetic_autocorrelation_taxonomy_67columns_107samples_feb2021.csv")

#-----------------------------------------------------------------------------
#Phylogenetic anova

#individually-  running the function in a list of columns-
tree<-read.tree("1000bps_taxonomy_107taxonomy_taxa_tree.nwk")
functions<-read.csv("taxonomy_107samples_67taxa_feb2021.csv",row.names = 1,header=TRUE)
functions$X<-NULL
functions<-functions[match(tree$tip.label,rownames(functions)), ] #match the tree
functions <-functions+100 #add a constant to every value

sampleinfo<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv")

library(geomorph)
phyl.anova.func2_list<-function(filename){
filename<-as.data.frame(t(filename)) #convert row to column
filename$label<-rownames(filename)
geo_functions<-merge(filename,sampleinfo,by.x="label",by.y="termite_tree_labels")
geo_functions$diet_type_number4<-as.factor(geo_functions$diet_type_number4)
geo_functions[,2]<-as.numeric(geo_functions[,2])

gdf <- geomorph.data.frame(abundance = geo_functions[,2], diet = as.factor(geo_functions$diet_type_number4), phy = tree) #create a dataframe that contains the data and phylogeny. [according to google forum]

#group_name <- names(filename)[2] #get the correct column name. column 2 is the cazyme.

geomorph.lm<-procD.lm(abundance ~ diet, iter=9999,RRPP=TRUE,effect.type = "F", SS.type = "II" ,data=gdf) #SS.type="II" is used if the groups (i.e. diet/lineage) are of uneven sizes.
mod<-anova(geomorph.lm) #to examine if there is a phylogenetic correlation between "y" and "x"
mod2<-as.data.frame(mod$table)
mod2<-mod2[1,7]
# d is distance between vector lengths is the difference in the amount of shape change per unit of size. If d is 0.05 i.e., | 0.20 - 0.15 |.  Why do this?  Two vectors can be correlated in direction - the way shape changes with size - but one group can exhibit more shape change per unit size than the other
#D tells us the effect size of the standard deviation between two groups. Positive and above 0.2-0.3 means larger positive difference, while negative and below -0.2 to -0.3 means larger negative difference.

pw<-pairwise(geomorph.lm,groups = as.factor(gdf$diet))
pw_summary<-summary(pw,confidence=0.95,show.vectors = TRUE)
pw_summary2<-as.data.frame(pw_summary$summary.table)

pw_summary2$padjust<-p.adjust(pw_summary2$`Pr > d`,method="fdr") #fdr correction of p.value
pw_summary2$diets<-rownames(pw_summary2)

#convert the output to correct formats-
names<-colnames(geo_functions[2])
pw_summary2$genename<-rep(names,times=nrow(pw_summary2))
library(tidyr)
pw_summary3<-pw_summary2%>%dplyr::select("padjust","diets","genename")
pw_spread<-pw_summary3%>%spread(diets,padjust)
pw_spread2<-cbind(pw_spread,mod2)

return(pw_spread2)
}

functions_t<-as.data.frame(t(functions))
xy.list <-  split(functions_t, seq(nrow(functions_t))) #convert df to a list

taxonomy_geomorph<-lapply(xy.list,phyl.anova.func2_list) #apply on all columns 
 
taxonomy_geomorph_df<-do.call(rbind.data.frame,taxonomy_geomorph)
 

#find p.adjust for anova-
taxonomy_geomorph_df$padjust<-p.adjust(taxonomy_geomorph_df$mod2,method="fdr")
taxonomy_geomorph_df<-taxonomy_geomorph_df%>%mutate(significant_mod2=ifelse(mod2 <=0.05,"*",""))
taxonomy_geomorph_df<-taxonomy_geomorph_df%>%mutate(significant_anova=ifelse(padjust <=0.05,"*",""))

write.csv(taxonomy_geomorph_df,file="Phylogenetic_anova_withdiet_taxonomy_67columns_107samples_feb2021.csv")


```


##basic stats for taxonomy-
```{r}
##adding domain information to all the taxonomy files-
functions<-read.csv("taxonomy_107samples_67taxa_feb2021.csv",row.names = 1,header=TRUE)
taxa<-as.data.frame(colnames(functions))
colnames(taxa)<-c("taxa")
taxa$phyla<-unlist(lapply(strsplit(as.character(taxa$taxa),split="_c__"),"[",1))
taxa$phyla<-gsub("p__","",taxa$phyla)

alldomains<-read.csv("gtdb_bac120_ar122_domaininfo.csv")

taxa2<-merge(taxa,alldomains,by="phyla")


#How many phyla are present in the final dataset?
taxa2%>%select(phyla,domain)%>%group_by(phyla,domain)%>%count()
taxa2%>%select(phyla,domain)%>%group_by(phyla,domain)%>%filter(domain=="Bacteria")%>%count()


#----------------------------------------------
#Distribution of Archaea in all samples + Mimeutermes sample-
taxonomy_rawtpm<-read.csv("taxonomy_124samples_456taxa_rawTPM_feb2021.csv",row.names = 1)
taxonomy_rawtpm$X<-rownames(taxonomy_rawtpm)

library(tidyr)
taxonomy_long<-gather(taxonomy_rawtpm, taxonomy, tpm, 1:456, factor_key=TRUE)

taxonomy_long$phyla<-unlist(lapply(strsplit(as.character(taxonomy_long$taxonomy),split="_c__"),"[",1))
taxonomy_long$phyla<-gsub("p__","",taxonomy_long$phyla)

alldomains<-read.csv("gtdb_bac120_ar122_domaininfo.csv")

taxonomy_long<-merge(taxonomy_long,alldomains,by="phyla")
taxonomy_long$X.y<-NULL
taxonomy_long<-taxonomy_long%>%filter(tpm>0)

##how many archaeal families present in the rawTPM data in 124 samples-
taxonomy_long%>%select(phyla,taxonomy,domain)%>%filter(domain=="Archaea")%>%unique()

##how many archaeal families present in statistical data in 107 samples-
taxa2%>%select(phyla,taxa,domain)%>%group_by(phyla,taxa,domain)%>%filter(domain=="Archaea")%>%count()
#--------------------------------------------------------------------------------
##Overall relative abundance of Archaea-

library(data.table)
setDT(taxonomy_long)
taxonomy_long_domains<-taxonomy_long[, .(tpm = sum(tpm)), by = .(domain)]

taxonomy_long_domains[2,2]/sum(taxonomy_long_domains$tpm)*100

print(paste0("Proportion of Archaea in 124 metagenomes are ", taxonomy_long_domains[2,2]/sum(taxonomy_long_domains$tpm)*100 ,"%"))

#-------------------------------------------------------------------------------
##Which sample has the most relative abundance of archaea?

library(data.table)
setDT(taxonomy_long)
taxonomy_long_sampledomains<-taxonomy_long[, .(tpm = sum(tpm)), by = .(X.x,domain)]
taxonomy_long_sampledomains_archaea<-taxonomy_long_sampledomains%>%filter(domain=="Archaea")
max(taxonomy_long_sampledomains_archaea$tpm)
taxonomy_long_sampledomains_archaea%>%filter(tpm==max(taxonomy_long_sampledomains_archaea$tpm)) #extract the samplename with most archaeal TPM in it.

mimeutermes<-taxonomy_long_sampledomains%>%filter(X.x=="272_32-CIVT017-Nasutitermitinae-Mimeutermes_sorex")
mimeutermes[2,3]/sum(mimeutermes$tpm)*100


mimeutermes2<-taxonomy_long%>%filter(X.x %in% c("272_32-CIVT017-Nasutitermitinae-Mimeutermes_sorex"))
mimeutermes_archaea<-mimeutermes2%>%filter(domain=="Archaea")
print(paste0("Proportion of Archaea in Mimeutermes sample is ", sum(mimeutermes_archaea$tpm)/sum(mimeutermes$tpm)*100))
print(paste0("Proportion of Bathyarchaeia in Mimeutermes sample is ", mimeutermes_archaea[1,4]/sum(mimeutermes_archaea$tpm)*100))

#NOTE-Mimeutermes has the most no. of Archaeal relative abundance in all samples, but it accounts for 29.11% of total.
#---------------------------------------------------------------------------------------
##do all soil-feeders have higher Archaeal relative abundance than other diets?
sampleinfo<-read.csv("sample_names_ids_withotherfactors_tomedit.2.csv")
taxonomy_long_diet<-merge(taxonomy_long,sampleinfo,by.x="X.x",by.y="termite_tree_labels")

library(data.table)
setDT(taxonomy_long_diet)
taxonomy_long_dietdomains<-taxonomy_long_diet[, .(tpm = sum(tpm)), by = .(diet_type_number4,domain)]
taxonomy_long_dietdomains<-taxonomy_long_dietdomains%>%group_by(diet_type_number4)%>%mutate(overallsum=sum(tpm))
taxonomy_long_dietdomains$perc<-taxonomy_long_dietdomains$tpm/taxonomy_long_dietdomains$overallsum*100

taxonomy_long_dietdomains%>%filter(diet_type_number4=="3" & domain=="Archaea")

taxonomy_long_dietdomains%>%filter(diet_type_number4=="2" & domain=="Archaea") 

#NOTE-Archaea are more abundant in soil-feeding termites, followed by fungal-cultivating termites.
```




##16S and marker gene taxonomy comparison-
```{r}
ncbi<-read.csv("16S_newdata_89samples_408columns.csv") #OTUs summed at family level
markergene<-read.csv("taxonomy_124samples_456taxa_rawTPM_feb2021.csv")

#how many samples are in common between the two datasets?
ncbi_markers<-merge(ncbi,markergene,by.x="termite_samples","X") #72 termite samples in common.
rownames(ncbi_markers)<-ncbi_markers$termite_samples

#how many family-level taxonomy in common between the two datasets?
library(stringr)
taxa<-colnames(ncbi_markers)%>%as.data.frame()
colnames(taxa)<-c("family")
taxa_16s<-taxa%>%filter(!str_detect(family, "p__"))
taxa_16s$family<-gsub("_unclassified","",taxa_16s$family)
taxa_16s$taxonomyofinterest<-gsub("^.*_", "",taxa_16s$family)

taxa_markers<-taxa%>%filter(str_detect(family, "p__"))
taxa_markers$taxonomyofinterest<-gsub("^.*_", "",taxa_markers$family)

commontaxa<-merge(taxa_markers,taxa_16s,by="taxonomyofinterest") #141 family-level taxa in common.
commontaxa_vector<-c(as.vector(commontaxa$family.x),as.vector(commontaxa$family.y))

names<-names(ncbi_markers)[(names(ncbi_markers) %in% commontaxa_vector)]
ncbi_markers_subset <- ncbi_markers[, names]

write.csv(ncbi_markers_subset,file="taxonomy_16S_markers_rawcounts.csv")
#NOTE- we have 72 termite samples with 141 common family-level taxa.
#-------------------------------------------------------------------------------------------
##extract out the "others" column in each file-

#ncbi 16S-
ncbi_long<-gather(ncbi,taxa,count,3:409)
ncbi_long$X<-NULL
ncbi_long<-ncbi_long%>%mutate(presence=ifelse(taxa %in% names,"common","others"))

markers_long<-gather(markergene,taxa,count,2:457)
markers_long<-markers_long%>%mutate(presence=ifelse(taxa %in% names,"common","others"))


#sum by "others"
setDT(ncbi_long)
ncbi_long_others<-ncbi_long[, .(count = sum(count)), by = .(termite_samples,presence)]
ncbi_long_others<-ncbi_long_others%>%filter(presence=="others")


setDT(markers_long)
markers_long_others<-markers_long[, .(count = sum(count)), by = .(X,presence)]
markers_long_others<-markers_long_others%>%filter(presence=="others")
#-------------------------------------------------------------------------------------------
##join all the data in one-
ncbi_markers_subset<-read.csv("taxonomy_16S_markers_rawcounts.csv")
ncbi_markers_subset2<-merge(ncbi_markers_subset,ncbi_long_others,by.x="X",by.y="termite_samples")
ncbi_markers_subset3<-merge(ncbi_markers_subset2,markers_long_others,by="X")
ncbi_markers_subset3$presence.y<-NULL
ncbi_markers_subset3$presence.x<-NULL

write.csv(ncbi_markers_subset3,file="taxonomy_16S_markers_others_rawcounts.csv")
#-------------------------------------------------------------------------------------------
##CLR-transform the data + apply filters
ncbi_markers_subset3<-read.csv("taxonomy_16S_markers_others_rawcounts.csv",row.names=1)
rownames(ncbi_markers_subset3)<-ncbi_markers_subset3$X
ncbi_markers_subset3$X<-NULL

clr_transform<-function(filename_spread2){ 
library(propr)
filename_spread2[is.na(filename_spread2)] <-0 #convert NA to zero
#rownames(filename_spread2)<-filename_spread2$label
filename_spread2$label<-NULL
filename_spread2$file_name<-NULL
#add psuedo counts to account for missing and not-present data-
filename_spread2<-filename_spread2+0.65

filename_propr<-propr(filename_spread2,metric="rho",ivar="clr")

#creating a df with all the transformed pathways and metadata-
filename_propr_df<-filename_propr@logratio

return(filename_propr_df)
}

ncbi_markers_clr<-clr_transform(ncbi_markers_subset3)

above5perc_function<-function(filename){
filename$label<-NULL  
filename$samplename<-NULL  

filename2<-filename #for filtering 
filename2[filename2 > 0] <- 1 
filename2[filename2 <= 0] <-0
filename_10perc<-filename2[colSums(filename2) > 4]  #4/72=5%
columns<-colnames(filename_10perc) 

filename3<-filename%>%dplyr::select(columns)
return(filename3)
}

ncbi_markers_clr_5perc<-above5perc_function(ncbi_markers_clr)

#from clr transformed data-> re-select rows with more than 3 functions 
morethan3taxa_functions<-function(filename){
 
filename$label<-NULL 
filename$samplename<-NULL
filename2<-filename
filename2[filename2 > 0] <- 1
filename2[filename2 <= 0] <-0
filename_threeabove<-filename2[rowSums(filename2[,])>=3, ]
rows<-rownames(filename_threeabove)

filename3<-subset(filename, rownames(filename) %in% rows)

return(filename3)
}

ncbi_markers_clr_5perc_3above<-morethan3taxa_functions(ncbi_markers_clr_5perc)


#how many family-level taxa present?
taxa<-colnames(ncbi_markers_clr_5perc_3above)%>%as.data.frame()
colnames(taxa)<-c("family")
taxa_16s<-taxa%>%filter(!str_detect(family, "p__"))
#taxa_16s$family<-gsub("_unclassified","",taxa_16s$family)
taxa_16s$taxonomyofinterest<-gsub("^.*_", "",taxa_16s$family)

taxa_markers<-taxa%>%filter(str_detect(family, "p__"))
taxa_markers$taxonomyofinterest<-gsub("^.*_", "",taxa_markers$family)

commontaxa<-merge(taxa_markers,taxa_16s,by="taxonomyofinterest") #70 family-level taxa in common.
commontaxa_vector<-c(as.vector(commontaxa$family.x),as.vector(commontaxa$family.y),"count.x","count.y")

names<-names(ncbi_markers_clr_5perc_3above)[(names(ncbi_markers_clr_5perc_3above) %in% commontaxa_vector)]
ncbi_markers_clr_5perc_3above_subset <- ncbi_markers_clr_5perc_3above[, names]

write.csv(ncbi_markers_clr_5perc_3above_subset,file="taxonomy_16S_markers_others_clr.csv")

```







##basic stats on choosing the marker taxonomy vs contig taxonomy-
```{r}
##q1. Is there a correlation between no. of markers and their length?

allcogs<-read.csv("tpm_cogs_allsamples_feb2021_1000bpscontigs_100counts1tpm.csv")
library(dplyr)

allcogs2<-allcogs%>%select(gene_length,markers)

#get mean length per marker
library(data.table)
setDT(allcogs2)
allcogs_mean<-allcogs2[, .(mean_length = mean(gene_length)), by = .(markers)]

#get no. of markers annotated
allcogs_number<-allcogs2%>%select(markers)%>%group_by(markers)%>%count()

#join two tables-
allcogs_info<-merge(allcogs_mean,allcogs_number,by="markers")

##do the same for a)markers that have same taxonomy as contigs ("tpm_cogs_allsamples_feb2021_1000bpscontigs_100counts1pm_similartocontigs.csv"), b)markers that dont ("cog2_notgtdbcontigs.csv")

cogs_similar<-read.csv("tpm_cogs_allsamples_feb2021_1000bpscontigs_100counts1pm_similartocontigs.csv")
cogs_similar2<-cogs_similar%>%select(gene_length,markers)
setDT(cogs_similar2)
cogssimilar_mean<-cogs_similar2[, .(mean_length = mean(gene_length)), by = .(markers)]
cogssimilar_number<-cogs_similar2%>%select(markers)%>%group_by(markers)%>%count()
cogssimilar_info<-merge(cogssimilar_mean,cogssimilar_number,by="markers")

cogs_dif<-read.csv("cog2_notgtdbcontigs.csv")
cogs_dif2<-cogs_dif%>%select(gene_length,markers)
setDT(cogs_dif2)
cogsdif_mean<-cogs_dif2[, .(mean_length = mean(gene_length)), by = .(markers)]
cogsdif_number<-cogs_dif2%>%select(markers)%>%group_by(markers)%>%count()
cogsdif_info<-merge(cogsdif_mean,cogsdif_number,by="markers")


#correlations-
res_allcogs<-cor.test(allcogs_info$mean_length, allcogs_info$n, method="pearson")
#[1] p-value = 5.809e-11
res_cogssimilar<-cor.test(cogssimilar_info$mean_length, cogssimilar_info$n, method="pearson")
#[1] p-value = 7.257e-11
res_cogsdif<-cor.test(cogsdif_info$mean_length, cogsdif_info$n, method="pearson")
#[1] p-value = 2.826e-06

#------------------------------------------------------------------------------------------------------
#q2. Is taxonomy marker gene specific AND  do different files have different families?
allcogs_taxa<-allcogs%>%select(markers,family,prokTPM)
allcogs_similar_taxa<-cogs_similar%>%select(markers,family,prokTPM)
allcogs_dif_taxa<-cogs_dif%>%select(markers,family,prokTPM)

library(data.table)
setDT(allcogs_taxa)
allcogs_taxasum<-allcogs_taxa[, .(prokTPM = sum(prokTPM)), by = .(markers,family)]

setDT(allcogs_similar_taxa)
allcogs_similar_taxasum<-allcogs_similar_taxa[, .(prokTPM = sum(prokTPM)), by = .(markers,family)]

setDT(allcogs_dif_taxa)
allcogs_dif_taxasum<-allcogs_dif_taxa[, .(prokTPM = sum(prokTPM)), by = .(markers,family)]

#choosing a subset of taxa to visualize (based on phyl.anova results)
taxa_select<-c("p__Bacteroidota_c__Bacteroidia_o__Bacteroidales","p__Bacteroidota_c__Bacteroidia_o__Bacteroidales_f__Bacteroidaceae","p__Bacteroidota_c__Bacteroidia_o__Bacteroidales_f__Dysgonomonadaceae","p__Bacteroidota_c__Bacteroidia_o__Bacteroidales_f__Tannerellaceae","p__Bacteroidota_c__Bacteroidia_o__Bacteroidales_f__UBA932","p__Bacteroidota_c__Bacteroidia_o__Bacteroidales_f__vadinHA17","p__Campylobacterota_c__Campylobacteria_o__Campylobacterales_f__Campylobacteraceae","p__Crenarchaeota_c__Bathyarchaeia_o__B26.1_f__UBA233","p__Elusimicrobiota_c__Endomicrobia_o__Endomicrobiales_f__Endomicrobiaceae","p__Fibrobacterota_c__Chitinivibrionia_o__Chitinivibrionales_f__Chitinispirillaceae","p__Fibrobacterota_c__Fibrobacteria_o__Fibrobacterales_f__Fibrobacteraceae","p__Firmicutes_c__Bacilli_o__Bacillales_f__Amphibacillaceae","p__Firmicutes_c__Bacilli_o__Lactobacillales_f__Lactobacillaceae","p__Firmicutes_c__Bacilli_o__Mycoplasmatales_f__Mycoplasmoidaceae","p__Proteobacteria_c__Alphaproteobacteria_o__Caulobacterales_f__Hyphomonadaceae","p__Proteobacteria_c__Gammaproteobacteria_o__Burkholderiales_f__Burkholderiaceae","p__Proteobacteria_c__Gammaproteobacteria_o__Enterobacterales_f__Enterobacteriaceae","p__Spirochaetota_c__Spirochaetia_o__Sphaerochaetales_f__Sphaerochaetaceae","p__Spirochaetota_c__Spirochaetia_o__Treponematales_f__Treponemataceae_B","p__Spirochaetota_c__Spirochaetia_o__Spirochaetales_f__RBG.16.67.19")

allcogs_taxasum2<-allcogs_taxasum%>%filter(family %in% taxa_select)
allcogs_simialr_taxasum2<-allcogs_similar_taxasum%>%filter(family %in% taxa_select)
allcogs_dif_taxasum2<-allcogs_dif_taxasum%>%filter(family %in% taxa_select)


#summing all markers-
setDT(allcogs_taxa)
allcogs_onlytaxasum<-allcogs_taxa[, .(prokTPM = sum(prokTPM)), by = .(family)]
allcogs_onlytaxasum<-allcogs_onlytaxasum%>%filter(family %in% taxa_select)
allcogs_onlytaxasum$markers<-rep("allcogs",nrow(allcogs_onlytaxasum))

setDT(allcogs_similar_taxa)
allcogs_similar_onlytaxasum<-allcogs_similar_taxa[, .(prokTPM = sum(prokTPM)), by = .(family)]
allcogs_similar_onlytaxasum<-allcogs_similar_onlytaxasum%>%filter(family %in% taxa_select)
allcogs_similar_onlytaxasum$markers<-rep("allcogs_similar",nrow(allcogs_similar_onlytaxasum))

setDT(allcogs_dif_taxa)
allcogs_dif_onlytaxasum<-allcogs_dif_taxa[, .(prokTPM = sum(prokTPM)), by = .(family)]
allcogs_dif_onlytaxasum<-allcogs_dif_onlytaxasum%>%filter(family %in% taxa_select)
allcogs_dif_onlytaxasum$markers<-rep("allcogs_dif",nrow(allcogs_dif_onlytaxasum))

#plot-
library(ggplot2)
#colors used for microbial taxa-based on https://github.com/katiejolly/nationalparkcolors/blob/master/R/colors.R
##BASED ON MAG TREE JAN2021-
allcolors<-c(p__Bacteroidota_c__Bacteroidia_o__Bacteroidales="#9B593F",p__Bacteroidota_c__Bacteroidia_o__Bacteroidales_f__Bacteroidaceae="#2c5aa0",p__Bacteroidota_c__Bacteroidia_o__Bacteroidales_f__Dysgonomonadaceae="#00aa88",p__Bacteroidota_c__Bacteroidia_o__Bacteroidales_f__Tannerellaceae="#7c916f",p__Bacteroidota_c__Bacteroidia_o__Bacteroidales_f__UBA932="#E39B38",p__Bacteroidota_c__Bacteroidia_o__Bacteroidales_f__vadinHA17="#8C9D57",p__Campylobacterota_c__Campylobacteria_o__Campylobacterales_f__Campylobacteraceae="#4E5462",p__Crenarchaeota_c__Bathyarchaeia_o__B26.1_f__UBA233="#502d16",p__Elusimicrobiota_c__Endomicrobia_o__Endomicrobiales_f__Endomicrobiaceae="#8C9D57",p__Fibrobacterota_c__Chitinivibrionia_o__Chitinivibrionales_f__Chitinispirillaceae="#916f7c",p__Fibrobacterota_c__Fibrobacteria_o__Fibrobacterales_f__Fibrobacteraceae="#2E8289",p__Firmicutes_c__Bacilli_o__Bacillales_f__Amphibacillaceae="#6E687E",p__Firmicutes_c__Bacilli_o__Lactobacillales_f__Lactobacillaceae="#d45500",p__Firmicutes_c__Bacilli_o__Mycoplasmatales_f__Mycoplasmoidaceae="#A1045A",p__Proteobacteria_c__Alphaproteobacteria_o__Caulobacterales_f__Hyphomonadaceae="#A8CDEC",p__Proteobacteria_c__Gammaproteobacteria_o__Burkholderiales_f__Burkholderiaceae="#E79498",p__Proteobacteria_c__Gammaproteobacteria_o__Enterobacterales_f__Enterobacteriaceae="#FAB57C",p__Spirochaetota_c__Spirochaetia_o__Sphaerochaetales_f__Sphaerochaetaceae="#F6955E",p__Spirochaetota_c__Spirochaetia_o__Treponematales_f__Treponemataceae_B="#F1E3B6",p__Spirochaetota_c__Spirochaetia_o__Spirochaetales_f__RBG.16.67.19="black")

#overall theme-
addline_format <- function(x,...){
    gsub('\\s','\n',x)
}

theme_set(theme_minimal() + theme(legend.position = "right", axis.title =  element_text(size = 9), axis.text =  element_text(size = 9), strip.text = element_text(size = 9)))


p1<-allcogs_onlytaxasum%>%ggplot(aes(fill=family, y=prokTPM, x=as.factor(markers))) + geom_bar(stat="identity",position="fill")+scale_fill_manual(values=allcolors)+labs(y="Relative abundance of microbial families in summed markers",x="all markers")+theme(plot.title = element_text(hjust = 0.5))

p2<-allcogs_similar_onlytaxasum%>%ggplot(aes(fill=family, y=prokTPM, x=as.factor(markers))) + geom_bar(stat="identity",position="fill")+scale_fill_manual(values=allcolors)+labs(y="Relative abundance of microbial families in summed markers",x="all markers")+theme(plot.title = element_text(hjust = 0.5))

p3<-allcogs_dif_onlytaxasum%>%ggplot(aes(fill=family, y=prokTPM, x=as.factor(markers))) + geom_bar(stat="identity",position="fill")+scale_fill_manual(values=allcolors)+labs(y="Relative abundance of microbial families in summed markers",x="all markers")+theme(plot.title = element_text(hjust = 0.5))


q<-allcogs_taxasum2%>%ggplot( aes(fill=family, y=prokTPM, x=as.factor(markers))) + geom_bar(stat="identity",position="fill")+scale_fill_manual(values=allcolors,na.value="black")+labs(y="Relative abundance of microbial families in individual markers",x="all markers",title="individual markers allcogs")+theme(plot.title = element_text(hjust = 0.5))

library(ggpubr)
figure1 <- ggarrange(p1, q,labels = c("A", "B"),ncol = 2, nrow = 1,common.legend = TRUE, legend = "right")

pdf(file="markergenes_feb2021_1.pdf",height=8.5,width=13)
figure1
dev.off()

figure2 <- ggarrange(p1, p2,p3,
                    labels = c("A", "B","C"),
                    ncol = 3, nrow = 1,
                    common.legend = TRUE, legend = "right")

pdf(file="markergenes_feb2021_2.pdf",height=8.5,width=13)
figure2
dev.off()

#NOTE- Even though there is a correlation in no. of markers and marker length, the represented families in "allcogs" file and "allcogs_similar" files are the same (with different TPM values).


#-----------------------------------------------------------------------------------------
#taxonomic profile of contig taxonomy file-


allcontigs<-read.csv("uniq-all-metagenome-annotation-taxonomy-1000bps_feb2021.txt")
allcontigs_sub<-allcontigs%>%select(V1.y,prokTPM)
allcontigs_sub$V1.y<-gsub("\\|","_",allcontigs_sub$V1.y)
allcontigs_sub$family<-gsub("_g__.*$","",allcontigs_sub$V1.y)

setDT(allcontigs_sub)
allcontigs_taxasum<-allcontigs_sub[, .(prokTPM = sum(prokTPM)), by = .(family)]
allcontigs_taxasum<-allcontigs_taxasum%>%filter(family %in% taxa_select)
allcontigs_taxasum$markers<-rep("allcontigs",nrow(allcontigs_taxasum))


r<-allcontigs_taxasum%>%ggplot(aes(fill=family, y=prokTPM, x=as.factor(markers))) + geom_bar(stat="identity",position="fill")+scale_fill_manual(values=allcolors)+labs(y="Relative abundance of microbial families in summedcontigs",x="all contigs")+theme(plot.title = element_text(hjust = 0.5))

figure3 <- ggarrange(p1, p2,p3,r,
                    labels = c("A", "B","C","D"),
                    ncol = 4, nrow = 1,
                    common.legend = TRUE, legend = "right")

pdf(file="markergenes_feb2021_3.pdf",height=8.5,width=13)
figure3
dev.off()

##NOTE-Relative abundance of top 20 microbial families in contig annotation, allcogs annotation and allcogs similar to contigs are almost similar. We can use allcogs file for taxonomic annotation.
```

